From d8762f031947378fbfe585f6d6284fb98daf4166 Mon Sep 17 00:00:00 2001
From: Tam Nguyen <tam.nguyen.xa@renesas.com>
Date: Wed, 28 May 2025 02:35:57 +0700
Subject: [PATCH 37/46] net: renesas: rswitch3: Add support for R-Car X5H
 (r8a78000) SoC on evaluation board

This patch adds support for the real R-Car X5H (r8a78000) SoC with
adaptations for the new hardware blocks and register maps introduced
in the Gen5 architecture. The implementation is based on the existing
Ethernet Switch2 driver for R-Car S4-8, with the following updates:

- Update driver header with SPDX identifier and description
- Refine condition for setting MIOC configuration
- Add hardcoded module clock initialization for r8a78000
- Support using different ETHA channels for data transfer and MDIO
- Rename functions from "rswitch_" to "rsw3_" for naming consistency
- Remove runtime ETHA channel reconfiguration logic

Signed-off-by: Tam Nguyen <tam.nguyen.xa@renesas.com>
Signed-off-by: Phong Hoang <phong.hoang.wz@renesas.com>
Signed-off-by: Thanh Quan <thanh.quan.xn@renesas.com>
Signed-off-by: Hai Pham <hai.pham.ud@renesas.com>
---
 drivers/net/ethernet/renesas/rswitch3.c | 1137 ++++++++++-------------
 drivers/net/ethernet/renesas/rswitch3.h |  226 +++--
 2 files changed, 601 insertions(+), 762 deletions(-)

diff --git a/drivers/net/ethernet/renesas/rswitch3.c b/drivers/net/ethernet/renesas/rswitch3.c
index 92ed670556be..7aabda7cbe9b 100644
--- a/drivers/net/ethernet/renesas/rswitch3.c
+++ b/drivers/net/ethernet/renesas/rswitch3.c
@@ -1,9 +1,10 @@
-// SPDX-License-Identifier: GPL-2.0
-/* Renesas Ethernet Switch3 Device Driver
- * Based on the Ethernet Switch2 driver
- * from Linux Kernel v6.11-rc2
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Driver for Renesas Ethernet Switch3.
+ * The implementation is based on the existing Ethernet
+ * Switch2 driver for R-Car S4-8
  *
- * Copyright (C) 2024 Renesas Electronics Corporation
+ * Copyright (C) 2025 Renesas Electronics Corporation
  */
 
 #include <linux/clk.h>
@@ -32,12 +33,12 @@
 
 #include "rswitch3.h"
 
-static int rswitch_reg_wait(void __iomem *addr, u32 offs, u32 mask, u32 expected)
+static int rsw3_reg_wait(void __iomem *addr, u32 offs, u32 mask, u32 expected)
 {
 	u32 val;
 
 	if (!addr) {
-		pr_err("rswitch_reg_wait: Invalid address\n");
+		pr_err("rsw3_reg_wait: Invalid address\n");
 		return -EINVAL;
 	}
 
@@ -45,23 +46,23 @@ static int rswitch_reg_wait(void __iomem *addr, u32 offs, u32 mask, u32 expected
 										1, RSWITCH3_TIMEOUT_US);
 
 	if (ret) {
-		pr_err("rswitch_reg_wait: Timeout waiting for addr: 0x%p, offset: 0x%x, mask: 0x%x, expected: 0x%x\n",
+		pr_err("rsw3_reg_wait: Timeout waiting for addr: 0x%p, offset: 0x%x, mask: 0x%x, expected: 0x%x\n",
 			   addr, offs, mask, expected);
 	} else {
-		pr_debug("rswitch_reg_wait: Success addr: 0x%p, offset: 0x%x, val: 0x%x\n",
+		pr_debug("rsw3_reg_wait: Success addr: 0x%p, offset: 0x%x, val: 0x%x\n",
 				addr, offs, val);
 	}
 
 	return ret;
 }
 
-static void rswitch_modify(void __iomem *addr, enum rswitch_reg reg, u32 clear, u32 set)
+static void rsw3_modify(void __iomem *addr, enum rsw3_reg reg, u32 clear, u32 set)
 {
 	iowrite32((ioread32(addr + reg) & ~clear) | set, addr + reg);
 }
 
 /* Common Agent block (COMA) */
-static void rswitch_reset(struct rswitch_private *priv)
+static void rsw3_reset(struct rsw3_private *priv)
 {
 	if (!parallel_mode) {
 		iowrite32(RRC_RR, priv->addr + RRC);
@@ -95,19 +96,17 @@ static void rswitch_reset(struct rswitch_private *priv)
 	}
 }
 
-static void rswitch_clock_enable(struct rswitch_private *priv)
+static void rsw3_clock_enable(struct rsw3_private *priv)
 {
 	iowrite32(RCEC_ACE_DEFAULT | RCEC_RCE, priv->addr + RCEC);
-	pr_debug("rswitch: Clock enabled at addr: 0x%p\n", priv->addr);
 }
 
-static void rswitch_clock_disable(struct rswitch_private *priv)
+static void rsw3_clock_disable(struct rsw3_private *priv)
 {
 	iowrite32(RCDC_RCD, priv->addr + RCDC);
-	pr_debug("rswitch: Clock disabled at addr: 0x%p\n", priv->addr);
 }
 
-static bool rswitch_agent_clock_is_enabled(void __iomem *coma_addr,
+static bool rsw3_agent_clock_is_enabled(void __iomem *coma_addr,
 					   unsigned int port)
 {
 	u32 val = ioread32(coma_addr + RCEC);
@@ -120,7 +119,7 @@ static bool rswitch_agent_clock_is_enabled(void __iomem *coma_addr,
 	return !!(val & BIT(port));
 }
 
-static void rswitch_agent_clock_ctrl(void __iomem *coma_addr, unsigned int port,
+static void rsw3_agent_clock_ctrl(void __iomem *coma_addr, unsigned int port,
 				     int enable)
 {
 	u32 val;
@@ -128,15 +127,13 @@ static void rswitch_agent_clock_ctrl(void __iomem *coma_addr, unsigned int port,
 	if (enable) {
 		val = ioread32(coma_addr + RCEC);
 		iowrite32(val | RCEC_RCE | BIT(port), coma_addr + RCEC);
-		pr_debug("rswitch: Enabled agent clock for port %d\n", port);
 	} else {
 		val = ioread32(coma_addr + RCDC);
 		iowrite32(val | BIT(port), coma_addr + RCDC);
-		pr_debug("rswitch: Disabled agent clock for port %d\n", port);
 	}
 }
 
-static int rswitch_bpool_config(struct rswitch_private *priv)
+static int rsw3_bpool_config(struct rsw3_private *priv)
 {
 	u32 val;
 	int ret;
@@ -147,22 +144,22 @@ static int rswitch_bpool_config(struct rswitch_private *priv)
 
 	iowrite32(CABPIRM_BPIOG, priv->addr + CABPIRM);
 
-	ret = rswitch_reg_wait(priv->addr, CABPIRM, CABPIRM_BPR, CABPIRM_BPR);
+	ret = rsw3_reg_wait(priv->addr, CABPIRM, CABPIRM_BPR, CABPIRM_BPR);
 	if (ret) {
-		pr_err("rswitch_bpool_config: Timeout waiting for buffer pool configuration\n");
+		pr_err("rsw3_bpool_config: Timeout waiting for buffer pool configuration\n");
 		return ret;
 	}
 
 	return 0;
 }
 
-static void rswitch_coma_init(struct rswitch_private *priv)
+static void rsw3_coma_init(struct rsw3_private *priv)
 {
 	iowrite32(CABPPFLC_INIT_VALUE, priv->addr + CABPPFLC0);
 }
 
 /* R-Switch-3 block (TOP) */
-static void rswitch_top_init(struct rswitch_private *priv)
+static void rsw3_top_init(struct rsw3_private *priv)
 {
 	unsigned int i;
 
@@ -171,7 +168,7 @@ static void rswitch_top_init(struct rswitch_private *priv)
 }
 
 /* Forwarding engine block (MFWD) */
-static void rswitch_fwd_init(struct rswitch_private *priv)
+static void rsw3_fwd_init(struct rsw3_private *priv)
 {
 	unsigned int i;
 
@@ -182,6 +179,9 @@ static void rswitch_fwd_init(struct rswitch_private *priv)
 	}
 
 	for (i = 0; i < rswitch3_num_ports; i++) {
+		if (priv->rdev[i]->disabled)
+			continue;
+
 		iowrite32(priv->rdev[i]->rx_queue->index,
 			  priv->addr + FWPBFCSDC(GWCA_INDEX, i));
 		iowrite32(BIT(priv->gwca.index), priv->addr + FWPBFC(i));
@@ -194,47 +194,40 @@ static void rswitch_fwd_init(struct rswitch_private *priv)
 	iowrite32(GENMASK(rswitch3_num_ports - 1, 0), priv->addr + FWPBFC(priv->gwca.index));
 }
 
-/* gPTP timer (gPTP) */
-static void rswitch_get_timestamp(struct rswitch_private *priv,
-				  struct timespec64 *ts)
-{
-	priv->ptp_priv->info.gettime64(&priv->ptp_priv->info, ts);
-}
-
 /* Gateway CPU agent block (GWCA) */
-static int rswitch_gwca_change_mode(struct rswitch_private *priv,
-				    enum rswitch_gwca_mode mode)
+static int rsw3_gwca_change_mode(struct rsw3_private *priv,
+				    enum rsw3_gwca_mode mode)
 {
 	int ret;
 
-	if (!rswitch_agent_clock_is_enabled(priv->addr, priv->gwca.index))
-		rswitch_agent_clock_ctrl(priv->addr, priv->gwca.index, 1);
+	if (!rsw3_agent_clock_is_enabled(priv->addr, priv->gwca.index))
+		rsw3_agent_clock_ctrl(priv->addr, priv->gwca.index, 1);
 
 	iowrite32(mode, priv->addr + GWMC);
 
-	ret = rswitch_reg_wait(priv->addr, GWMS, GWMS_OPS_MASK, mode);
+	ret = rsw3_reg_wait(priv->addr, GWMS, GWMS_OPS_MASK, mode);
 
 	if (mode == GWMC_OPC_DISABLE)
-		rswitch_agent_clock_ctrl(priv->addr, priv->gwca.index, 0);
+		rsw3_agent_clock_ctrl(priv->addr, priv->gwca.index, 0);
 
 	return ret;
 }
 
-static int rswitch_gwca_mcast_table_reset(struct rswitch_private *priv)
+static int rsw3_gwca_mcast_table_reset(struct rsw3_private *priv)
 {
 	iowrite32(GWMTIRM_MTIOG, priv->addr + GWMTIRM);
 
-	return rswitch_reg_wait(priv->addr, GWMTIRM, GWMTIRM_MTR, GWMTIRM_MTR);
+	return rsw3_reg_wait(priv->addr, GWMTIRM, GWMTIRM_MTR, GWMTIRM_MTR);
 }
 
-static int rswitch_gwca_axi_ram_reset(struct rswitch_private *priv)
+static int rsw3_gwca_axi_ram_reset(struct rsw3_private *priv)
 {
 	iowrite32(GWARIRM_ARIOG, priv->addr + GWARIRM);
 
-	return rswitch_reg_wait(priv->addr, GWARIRM, GWARIRM_ARR, GWARIRM_ARR);
+	return rsw3_reg_wait(priv->addr, GWARIRM, GWARIRM_ARR, GWARIRM_ARR);
 }
 
-static bool rswitch_is_any_data_irq(struct rswitch_private *priv, u32 *dis, bool tx)
+static bool rsw3_is_any_data_irq(struct rsw3_private *priv, u32 *dis, bool tx)
 {
 	u32 *mask = tx ? priv->gwca.tx_irq_bits : priv->gwca.rx_irq_bits;
 	unsigned int i;
@@ -247,7 +240,7 @@ static bool rswitch_is_any_data_irq(struct rswitch_private *priv, u32 *dis, bool
 	return false;
 }
 
-static void rswitch_get_data_irq_status(struct rswitch_private *priv, u32 *dis)
+static void rsw3_get_data_irq_status(struct rsw3_private *priv, u32 *dis)
 {
 	int i;
 
@@ -255,7 +248,7 @@ static void rswitch_get_data_irq_status(struct rswitch_private *priv, u32 *dis)
 		dis[i] = ioread32(priv->addr + GWDIS0 + i * 0x10);
 }
 
-static void rswitch_enadis_data_irq(struct rswitch_private *priv, unsigned int index, bool enable)
+static void rsw3_enadis_data_irq(struct rsw3_private *priv, unsigned int index, bool enable)
 {
 	u32 offs = (enable ? GWDIE0 : GWDID0) + (index / 32) * 0x10;
 	u32 tmp = 0;
@@ -267,15 +260,15 @@ static void rswitch_enadis_data_irq(struct rswitch_private *priv, unsigned int i
 	iowrite32(BIT(index % 32) | tmp, priv->addr + offs);
 }
 
-static void rswitch_ack_data_irq(struct rswitch_private *priv, int index)
+static void rsw3_ack_data_irq(struct rsw3_private *priv, int index)
 {
 	u32 offs = GWDIS0 + (index / 32) * 0x10;
 
 	iowrite32(BIT(index % 32), priv->addr + offs);
 }
 
-static unsigned int rswitch_next_queue_index(struct rswitch_gwca_queue *gq,
-											 bool cur, unsigned int num)
+static unsigned int rsw3_next_queue_index(struct rsw3_gwca_queue *gq,
+					  bool cur, unsigned int num)
 {
 	unsigned int index = cur ? gq->cur : gq->dirty;
 
@@ -287,7 +280,7 @@ static unsigned int rswitch_next_queue_index(struct rswitch_gwca_queue *gq,
 	return index;
 }
 
-static unsigned int rswitch_get_num_cur_queues(struct rswitch_gwca_queue *gq)
+static unsigned int rsw3_get_num_cur_queues(struct rsw3_gwca_queue *gq)
 {
 	if (gq->cur >= gq->dirty)
 		return gq->cur - gq->dirty;
@@ -295,9 +288,9 @@ static unsigned int rswitch_get_num_cur_queues(struct rswitch_gwca_queue *gq)
 		return gq->ring_size - gq->dirty + gq->cur;
 }
 
-static bool rswitch_is_queue_rxed(struct rswitch_gwca_queue *gq)
+static bool rsw3_is_queue_rxed(struct rsw3_gwca_queue *gq)
 {
-	struct rswitch_ext_ts_desc *desc = &gq->rx_ring[gq->dirty];
+	struct rsw3_ext_ts_desc *desc = &gq->rx_ring[gq->dirty];
 
 	if ((desc->desc.die_dt & DT_MASK) != DT_FEMPTY)
 		return true;
@@ -305,9 +298,8 @@ static bool rswitch_is_queue_rxed(struct rswitch_gwca_queue *gq)
 	return false;
 }
 
-static int rswitch_gwca_queue_alloc_rx_buf(struct rswitch_gwca_queue *gq,
-										   unsigned int start_index,
-										   unsigned int num)
+static int rsw3_gwca_queue_alloc_rx_buf(struct rsw3_gwca_queue *gq,
+					   unsigned int start_index, unsigned int num)
 {
 	unsigned int i, index;
 
@@ -332,14 +324,15 @@ static int rswitch_gwca_queue_alloc_rx_buf(struct rswitch_gwca_queue *gq,
 	return -ENOMEM;
 }
 
-static void rswitch_gwca_queue_free(struct net_device *ndev,
-									struct rswitch_gwca_queue *gq)
+static void rsw3_gwca_queue_free(struct net_device *ndev,
+				   struct rsw3_gwca_queue *gq)
 {
 	unsigned int i;
 
+	return;
 	if (!gq->dir_tx) {
 		dma_free_coherent(ndev->dev.parent,
-				  sizeof(struct rswitch_ext_ts_desc) *
+				  sizeof(struct rsw3_ext_ts_desc) *
 				  (gq->ring_size + 1), gq->rx_ring, gq->ring_dma);
 		gq->rx_ring = NULL;
 
@@ -349,7 +342,7 @@ static void rswitch_gwca_queue_free(struct net_device *ndev,
 		gq->rx_bufs = NULL;
 	} else {
 		dma_free_coherent(ndev->dev.parent,
-				  sizeof(struct rswitch_ext_desc) *
+				  sizeof(struct rsw3_ext_desc) *
 				  (gq->ring_size + 1), gq->tx_ring, gq->ring_dma);
 		gq->tx_ring = NULL;
 		kfree(gq->skbs);
@@ -359,10 +352,10 @@ static void rswitch_gwca_queue_free(struct net_device *ndev,
 	}
 }
 
-static int rswitch_gwca_queue_alloc(struct net_device *ndev,
-									struct rswitch_private *priv,
-									struct rswitch_gwca_queue *gq,
-									bool dir_tx, unsigned int ring_size)
+static int rsw3_gwca_queue_alloc(struct net_device *ndev,
+				   struct rsw3_private *priv,
+				   struct rsw3_gwca_queue *gq,
+				   bool dir_tx, unsigned int ring_size)
 {
 	unsigned int i, bit;
 
@@ -374,11 +367,11 @@ static int rswitch_gwca_queue_alloc(struct net_device *ndev,
 		gq->rx_bufs = kcalloc(gq->ring_size, sizeof(*gq->rx_bufs), GFP_KERNEL);
 		if (!gq->rx_bufs)
 			return -ENOMEM;
-		if (rswitch_gwca_queue_alloc_rx_buf(gq, 0, gq->ring_size) < 0)
+		if (rsw3_gwca_queue_alloc_rx_buf(gq, 0, gq->ring_size) < 0)
 			goto out;
 
 		gq->rx_ring = dma_alloc_coherent(ndev->dev.parent,
-						 sizeof(struct rswitch_ext_ts_desc) *
+						 sizeof(struct rsw3_ext_ts_desc) *
 						 (gq->ring_size + 1), &gq->ring_dma, GFP_KERNEL);
 	} else {
 		gq->skbs = kcalloc(gq->ring_size, sizeof(*gq->skbs), GFP_KERNEL);
@@ -388,7 +381,7 @@ static int rswitch_gwca_queue_alloc(struct net_device *ndev,
 		if (!gq->unmap_addrs)
 			goto out;
 		gq->tx_ring = dma_alloc_coherent(ndev->dev.parent,
-						 sizeof(struct rswitch_ext_desc) *
+						 sizeof(struct rsw3_ext_desc) *
 						 (gq->ring_size + 1), &gq->ring_dma, GFP_KERNEL);
 	}
 
@@ -405,29 +398,29 @@ static int rswitch_gwca_queue_alloc(struct net_device *ndev,
 	return 0;
 
 out:
-	rswitch_gwca_queue_free(ndev, gq);
+	rsw3_gwca_queue_free(ndev, gq);
 
 	return -ENOMEM;
 }
 
-static void rswitch_desc_set_dptr(struct rswitch_desc *desc, dma_addr_t addr)
+static void rsw3_desc_set_dptr(struct rsw3_desc *desc, dma_addr_t addr)
 {
 	desc->dptrl = cpu_to_le32(lower_32_bits(addr));
 	desc->dptrh = upper_32_bits(addr) & 0xff;
 }
 
-static dma_addr_t rswitch_desc_get_dptr(const struct rswitch_desc *desc)
+static dma_addr_t rsw3_desc_get_dptr(const struct rsw3_desc *desc)
 {
 	return __le32_to_cpu(desc->dptrl) | (u64)(desc->dptrh) << 32;
 }
 
-static int rswitch_gwca_queue_format(struct net_device *ndev,
-									struct rswitch_private *priv,
-									struct rswitch_gwca_queue *gq)
+static int rsw3_gwca_queue_format(struct net_device *ndev,
+				    struct rsw3_private *priv,
+				    struct rsw3_gwca_queue *gq)
 {
-	unsigned int ring_size = sizeof(struct rswitch_ext_desc) * gq->ring_size;
-	struct rswitch_ext_desc *desc;
-	struct rswitch_desc *linkfix;
+	unsigned int ring_size = sizeof(struct rsw3_ext_desc) * gq->ring_size;
+	struct rsw3_ext_desc *desc;
+	struct rsw3_desc *linkfix;
 	dma_addr_t dma_addr;
 	unsigned int i;
 
@@ -442,28 +435,31 @@ static int rswitch_gwca_queue_format(struct net_device *ndev,
 				goto err;
 
 			desc->desc.info_ds = cpu_to_le16(RSWITCH3_DESC_BUF_SIZE);
-			rswitch_desc_set_dptr(&desc->desc, dma_addr);
+			rsw3_desc_set_dptr(&desc->desc, dma_addr);
 			desc->desc.die_dt = DT_FEMPTY | DIE;
 		} else {
 			desc->desc.die_dt = DT_EEMPTY | DIE;
 		}
 	}
-	rswitch_desc_set_dptr(&desc->desc, gq->ring_dma);
+	rsw3_desc_set_dptr(&desc->desc, gq->ring_dma);
 	desc->desc.die_dt = DT_LINKFIX;
 
 	linkfix = &priv->gwca.linkfix_table[gq->index];
 	linkfix->die_dt = DT_LINKFIX;
-	rswitch_desc_set_dptr(linkfix, gq->ring_dma);
+	rsw3_desc_set_dptr(linkfix, gq->ring_dma);
 
 	iowrite32(GWDCC_BALR | (gq->dir_tx ? GWDCC_DCP(GWCA_IPV_NUM) | GWDCC_DQT : 0) | GWDCC_EDE,
 		  priv->addr + GWDCC_OFFS(gq->index));
 
+	/* Enable Under Switch Minimum Frame Size Padding */
+	iowrite32(GWCKSC_USMFSPE, priv->addr + GWCKSC);
+
 	return 0;
 
 err:
 	if (!gq->dir_tx) {
 		for (desc = gq->tx_ring; i-- > 0; desc++) {
-			dma_addr = rswitch_desc_get_dptr(&desc->desc);
+			dma_addr = rsw3_desc_get_dptr(&desc->desc);
 			dma_unmap_single(ndev->dev.parent, dma_addr,
 					 RSWITCH3_MAP_BUF_SIZE, DMA_FROM_DEVICE);
 		}
@@ -472,13 +468,13 @@ static int rswitch_gwca_queue_format(struct net_device *ndev,
 	return -ENOMEM;
 }
 
-static int rswitch_gwca_queue_ext_ts_fill(struct net_device *ndev,
-					  struct rswitch_gwca_queue *gq,
+static int rsw3_gwca_queue_ext_ts_fill(struct net_device *ndev,
+					  struct rsw3_gwca_queue *gq,
 					  unsigned int start_index,
 					  unsigned int num)
 {
-	struct rswitch_device *rdev = netdev_priv(ndev);
-	struct rswitch_ext_ts_desc *desc;
+	struct rsw3_device *rdev = netdev_priv(ndev);
+	struct rsw3_ext_ts_desc *desc;
 	unsigned int i, index;
 	dma_addr_t dma_addr;
 
@@ -494,7 +490,7 @@ static int rswitch_gwca_queue_ext_ts_fill(struct net_device *ndev,
 				goto err;
 
 			desc->desc.info_ds = cpu_to_le16(RSWITCH3_DESC_BUF_SIZE);
-			rswitch_desc_set_dptr(&desc->desc, dma_addr);
+			rsw3_desc_set_dptr(&desc->desc, dma_addr);
 			dma_wmb();
 			desc->desc.die_dt = DT_FEMPTY | DIE;
 			desc->info1 = cpu_to_le64(INFO1_SPN(rdev->etha->index));
@@ -510,7 +506,7 @@ static int rswitch_gwca_queue_ext_ts_fill(struct net_device *ndev,
 		for (; i-- > 0; ) {
 			index = (i + start_index) % gq->ring_size;
 			desc = &gq->rx_ring[index];
-			dma_addr = rswitch_desc_get_dptr(&desc->desc);
+			dma_addr = rsw3_desc_get_dptr(&desc->desc);
 			dma_unmap_single(ndev->dev.parent, dma_addr,
 					 RSWITCH3_MAP_BUF_SIZE, DMA_FROM_DEVICE);
 		}
@@ -519,42 +515,45 @@ static int rswitch_gwca_queue_ext_ts_fill(struct net_device *ndev,
 	return -ENOMEM;
 }
 
-static int rswitch_gwca_queue_ext_ts_format(struct net_device *ndev,
-					    struct rswitch_private *priv,
-					    struct rswitch_gwca_queue *gq)
+static int rsw3_gwca_queue_ext_ts_format(struct net_device *ndev,
+					    struct rsw3_private *priv,
+					    struct rsw3_gwca_queue *gq)
 {
-	unsigned int ring_size = sizeof(struct rswitch_ext_ts_desc) * gq->ring_size;
-	struct rswitch_ext_ts_desc *desc;
-	struct rswitch_desc *linkfix;
+	unsigned int ring_size = sizeof(struct rsw3_ext_ts_desc) * gq->ring_size;
+	struct rsw3_ext_ts_desc *desc;
+	struct rsw3_desc *linkfix;
 	int err;
 
 	memset(gq->rx_ring, 0, ring_size);
-	err = rswitch_gwca_queue_ext_ts_fill(ndev, gq, 0, gq->ring_size);
+	err = rsw3_gwca_queue_ext_ts_fill(ndev, gq, 0, gq->ring_size);
 	if (err < 0)
 		return err;
 
 	desc = &gq->rx_ring[gq->ring_size];	/* Last */
-	rswitch_desc_set_dptr(&desc->desc, gq->ring_dma);
+	rsw3_desc_set_dptr(&desc->desc, gq->ring_dma);
 	desc->desc.die_dt = DT_LINKFIX;
 
 	linkfix = &priv->gwca.linkfix_table[gq->index];
 	linkfix->die_dt = DT_LINKFIX;
-	rswitch_desc_set_dptr(linkfix, gq->ring_dma);
+	rsw3_desc_set_dptr(linkfix, gq->ring_dma);
 
 	iowrite32(GWDCC_BALR | (gq->dir_tx ? GWDCC_DCP(GWCA_IPV_NUM) | GWDCC_DQT : 0) |
 		  GWDCC_ETS | GWDCC_EDE,
 		  priv->addr + GWDCC_OFFS(gq->index));
 
+	/* Enable Under Switch Minimum Frame Size Padding */
+	iowrite32(GWCKSC_USMFSPE, priv->addr + GWCKSC);
+
 	return 0;
 }
 
-static int rswitch_gwca_linkfix_alloc(struct rswitch_private *priv)
+static int rsw3_gwca_linkfix_alloc(struct rsw3_private *priv)
 {
 	unsigned int i, num_queues = priv->gwca.num_queues;
-	struct rswitch_gwca *gwca = &priv->gwca;
+	struct rsw3_gwca *gwca = &priv->gwca;
 	struct device *dev = &priv->pdev->dev;
 
-	gwca->linkfix_table_size = sizeof(struct rswitch_desc) * num_queues;
+	gwca->linkfix_table_size = sizeof(struct rsw3_desc) * num_queues;
 	gwca->linkfix_table = dma_alloc_coherent(dev, gwca->linkfix_table_size,
 						 &gwca->linkfix_table_dma, GFP_KERNEL);
 	if (!gwca->linkfix_table)
@@ -565,9 +564,9 @@ static int rswitch_gwca_linkfix_alloc(struct rswitch_private *priv)
 	return 0;
 }
 
-static void rswitch_gwca_linkfix_free(struct rswitch_private *priv)
+static void rsw3_gwca_linkfix_free(struct rsw3_private *priv)
 {
-	struct rswitch_gwca *gwca = &priv->gwca;
+	struct rsw3_gwca *gwca = &priv->gwca;
 
 	if (gwca->linkfix_table)
 		dma_free_coherent(&priv->pdev->dev, gwca->linkfix_table_size,
@@ -575,9 +574,9 @@ static void rswitch_gwca_linkfix_free(struct rswitch_private *priv)
 	gwca->linkfix_table = NULL;
 }
 
-static struct rswitch_gwca_queue *rswitch_gwca_get(struct rswitch_private *priv)
+static struct rsw3_gwca_queue *rsw3_gwca_get(struct rsw3_private *priv)
 {
-	struct rswitch_gwca_queue *gq;
+	struct rsw3_gwca_queue *gq;
 	unsigned int index;
 
 	index = find_first_zero_bit(priv->gwca.used, priv->gwca.num_queues);
@@ -591,97 +590,97 @@ static struct rswitch_gwca_queue *rswitch_gwca_get(struct rswitch_private *priv)
 	return gq;
 }
 
-static void rswitch_gwca_put(struct rswitch_private *priv,
-			     struct rswitch_gwca_queue *gq)
+static void rsw3_gwca_put(struct rsw3_private *priv,
+			     struct rsw3_gwca_queue *gq)
 {
 	clear_bit(gq->index, priv->gwca.used);
 }
 
-static int rswitch_txdmac_alloc(struct net_device *ndev)
+static int rsw3_txdmac_alloc(struct net_device *ndev)
 {
-	struct rswitch_device *rdev = netdev_priv(ndev);
-	struct rswitch_private *priv = rdev->priv;
+	struct rsw3_device *rdev = netdev_priv(ndev);
+	struct rsw3_private *priv = rdev->priv;
 	int err;
 
-	rdev->tx_queue = rswitch_gwca_get(priv);
+	rdev->tx_queue = rsw3_gwca_get(priv);
 	if (!rdev->tx_queue)
 		return -EBUSY;
 
-	err = rswitch_gwca_queue_alloc(ndev, priv, rdev->tx_queue, true, TX_RING_SIZE);
+	err = rsw3_gwca_queue_alloc(ndev, priv, rdev->tx_queue, true, TX_RING_SIZE);
 	if (err < 0) {
-		rswitch_gwca_put(priv, rdev->tx_queue);
+		rsw3_gwca_put(priv, rdev->tx_queue);
 		return err;
 	}
 
 	return 0;
 }
 
-static void rswitch_txdmac_free(struct net_device *ndev)
+static void rsw3_txdmac_free(struct net_device *ndev)
 {
-	struct rswitch_device *rdev = netdev_priv(ndev);
+	struct rsw3_device *rdev = netdev_priv(ndev);
 
-	rswitch_gwca_queue_free(ndev, rdev->tx_queue);
-	rswitch_gwca_put(rdev->priv, rdev->tx_queue);
+	rsw3_gwca_queue_free(ndev, rdev->tx_queue);
+	rsw3_gwca_put(rdev->priv, rdev->tx_queue);
 }
 
-static int rswitch_txdmac_init(struct rswitch_private *priv, unsigned int index)
+static int rsw3_txdmac_init(struct rsw3_private *priv, unsigned int index)
 {
-	struct rswitch_device *rdev = priv->rdev[index];
+	struct rsw3_device *rdev = priv->rdev[index];
 
-	return rswitch_gwca_queue_format(rdev->ndev, priv, rdev->tx_queue);
+	return rsw3_gwca_queue_format(rdev->ndev, priv, rdev->tx_queue);
 }
 
-static int rswitch_rxdmac_alloc(struct net_device *ndev)
+static int rsw3_rxdmac_alloc(struct net_device *ndev)
 {
-	struct rswitch_device *rdev = netdev_priv(ndev);
-	struct rswitch_private *priv = rdev->priv;
+	struct rsw3_device *rdev = netdev_priv(ndev);
+	struct rsw3_private *priv = rdev->priv;
 	int err;
 
-	rdev->rx_queue = rswitch_gwca_get(priv);
+	rdev->rx_queue = rsw3_gwca_get(priv);
 	if (!rdev->rx_queue)
 		return -EBUSY;
 
-	err = rswitch_gwca_queue_alloc(ndev, priv, rdev->rx_queue, false, RX_RING_SIZE);
+	err = rsw3_gwca_queue_alloc(ndev, priv, rdev->rx_queue, false, RX_RING_SIZE);
 	if (err < 0) {
-		rswitch_gwca_put(priv, rdev->rx_queue);
+		rsw3_gwca_put(priv, rdev->rx_queue);
 		return err;
 	}
 
 	return 0;
 }
 
-static void rswitch_rxdmac_free(struct net_device *ndev)
+static void rsw3_rxdmac_free(struct net_device *ndev)
 {
-	struct rswitch_device *rdev = netdev_priv(ndev);
+	struct rsw3_device *rdev = netdev_priv(ndev);
 
-	rswitch_gwca_queue_free(ndev, rdev->rx_queue);
-	rswitch_gwca_put(rdev->priv, rdev->rx_queue);
+	rsw3_gwca_queue_free(ndev, rdev->rx_queue);
+	rsw3_gwca_put(rdev->priv, rdev->rx_queue);
 }
 
-static int rswitch_rxdmac_init(struct rswitch_private *priv, unsigned int index)
+static int rsw3_rxdmac_init(struct rsw3_private *priv, unsigned int index)
 {
-	struct rswitch_device *rdev = priv->rdev[index];
+	struct rsw3_device *rdev = priv->rdev[index];
 	struct net_device *ndev = rdev->ndev;
 
-	return rswitch_gwca_queue_ext_ts_format(ndev, priv, rdev->rx_queue);
+	return rsw3_gwca_queue_ext_ts_format(ndev, priv, rdev->rx_queue);
 }
 
-static int rswitch_gwca_hw_init(struct rswitch_private *priv)
+static int rsw3_gwca_hw_init(struct rsw3_private *priv)
 {
 	unsigned int i;
 	int err;
 
-	err = rswitch_gwca_change_mode(priv, GWMC_OPC_DISABLE);
+	err = rsw3_gwca_change_mode(priv, GWMC_OPC_DISABLE);
 	if (err < 0)
 		return err;
-	err = rswitch_gwca_change_mode(priv, GWMC_OPC_CONFIG);
+	err = rsw3_gwca_change_mode(priv, GWMC_OPC_CONFIG);
 	if (err < 0)
 		return err;
 
-	err = rswitch_gwca_mcast_table_reset(priv);
+	err = rsw3_gwca_mcast_table_reset(priv);
 	if (err < 0)
 		return err;
-	err = rswitch_gwca_axi_ram_reset(priv);
+	err = rsw3_gwca_axi_ram_reset(priv);
 	if (err < 0)
 		return err;
 
@@ -695,50 +694,52 @@ static int rswitch_gwca_hw_init(struct rswitch_private *priv)
 	iowrite32(GWTPC_PPPL(GWCA_IPV_NUM), priv->addr + GWTPC0);
 
 	for (i = 0; i < rswitch3_num_ports; i++) {
-		err = rswitch_rxdmac_init(priv, i);
+		if (priv->rdev[i]->disabled)
+			continue;
+		err = rsw3_rxdmac_init(priv, i);
 		if (err < 0)
 			return err;
-		err = rswitch_txdmac_init(priv, i);
+		err = rsw3_txdmac_init(priv, i);
 		if (err < 0)
 			return err;
 	}
 
-	err = rswitch_gwca_change_mode(priv, GWMC_OPC_DISABLE);
+	err = rsw3_gwca_change_mode(priv, GWMC_OPC_DISABLE);
 	if (err < 0)
 		return err;
-	return rswitch_gwca_change_mode(priv, GWMC_OPC_OPERATION);
+	return rsw3_gwca_change_mode(priv, GWMC_OPC_OPERATION);
 }
 
-static int rswitch_gwca_hw_deinit(struct rswitch_private *priv)
+static int rsw3_gwca_hw_deinit(struct rsw3_private *priv)
 {
 	int err;
 
-	err = rswitch_gwca_change_mode(priv, GWMC_OPC_DISABLE);
+	err = rsw3_gwca_change_mode(priv, GWMC_OPC_DISABLE);
 	if (err < 0)
 		return err;
-	err = rswitch_gwca_change_mode(priv, GWMC_OPC_RESET);
+	err = rsw3_gwca_change_mode(priv, GWMC_OPC_RESET);
 	if (err < 0)
 		return err;
 
-	return rswitch_gwca_change_mode(priv, GWMC_OPC_DISABLE);
+	return rsw3_gwca_change_mode(priv, GWMC_OPC_DISABLE);
 }
 
-static int rswitch_gwca_halt(struct rswitch_private *priv)
+static int rsw3_gwca_halt(struct rsw3_private *priv)
 {
 	int err;
 
 	priv->gwca_halt = true;
-	err = rswitch_gwca_hw_deinit(priv);
+	err = rsw3_gwca_hw_deinit(priv);
 	dev_err(&priv->pdev->dev, "halted (%d)\n", err);
 
 	return err;
 }
 
-static struct sk_buff *rswitch_rx_handle_desc(struct net_device *ndev,
-					      struct rswitch_gwca_queue *gq,
-					      struct rswitch_ext_ts_desc *desc)
+static struct sk_buff *rsw3_rx_handle_desc(struct net_device *ndev,
+					      struct rsw3_gwca_queue *gq,
+					      struct rsw3_ext_ts_desc *desc)
 {
-	dma_addr_t dma_addr = rswitch_desc_get_dptr(&desc->desc);
+	dma_addr_t dma_addr = rsw3_desc_get_dptr(&desc->desc);
 	u16 pkt_len = le16_to_cpu(desc->desc.info_ds) & RX_DS;
 	u8 die_dt = desc->desc.die_dt & DT_MASK;
 	struct sk_buff *skb = NULL;
@@ -808,15 +809,14 @@ static struct sk_buff *rswitch_rx_handle_desc(struct net_device *ndev,
 	return skb;
 }
 
-static bool rswitch_rx(struct net_device *ndev, int *quota)
+static bool rsw3_rx(struct net_device *ndev, int *quota)
 {
-	struct rswitch_device *rdev = netdev_priv(ndev);
-	struct rswitch_gwca_queue *gq = rdev->rx_queue;
-	struct rswitch_ext_ts_desc *desc;
+	struct rsw3_device *rdev = netdev_priv(ndev);
+	struct rsw3_gwca_queue *gq = rdev->rx_queue;
+	struct rsw3_ext_ts_desc *desc;
 	int limit, boguscnt, ret;
 	struct sk_buff *skb;
 	unsigned int num;
-	u32 get_ts;
 
 	if (*quota <= 0)
 		return true;
@@ -827,21 +827,10 @@ static bool rswitch_rx(struct net_device *ndev, int *quota)
 	desc = &gq->rx_ring[gq->cur];
 	while ((desc->desc.die_dt & DT_MASK) != DT_FEMPTY) {
 		dma_rmb();
-		skb = rswitch_rx_handle_desc(ndev, gq, desc);
+		skb = rsw3_rx_handle_desc(ndev, gq, desc);
 		if (!skb)
 			goto out;
 
-		get_ts = rdev->priv->ptp_priv->tstamp_rx_ctrl & RCAR_GEN4_RXTSTAMP_TYPE_V2_L2_EVENT;
-		if (get_ts) {
-			struct skb_shared_hwtstamps *shhwtstamps;
-			struct timespec64 ts;
-
-			shhwtstamps = skb_hwtstamps(skb);
-			memset(shhwtstamps, 0, sizeof(*shhwtstamps));
-			ts.tv_sec = __le32_to_cpu(desc->ts_sec);
-			ts.tv_nsec = __le32_to_cpu(desc->ts_nsec & cpu_to_le32(0x3fffffff));
-			shhwtstamps->hwtstamp = timespec64_to_ktime(ts);
-		}
 		skb->protocol = eth_type_trans(skb, ndev);
 		napi_gro_receive(&rdev->napi, skb);
 		rdev->ndev->stats.rx_packets++;
@@ -849,41 +838,41 @@ static bool rswitch_rx(struct net_device *ndev, int *quota)
 
 out:
 		gq->rx_bufs[gq->cur] = NULL;
-		gq->cur = rswitch_next_queue_index(gq, true, 1);
+		gq->cur = rsw3_next_queue_index(gq, true, 1);
 		desc = &gq->rx_ring[gq->cur];
 
 		if (--boguscnt <= 0)
 			break;
 	}
 
-	num = rswitch_get_num_cur_queues(gq);
-	ret = rswitch_gwca_queue_alloc_rx_buf(gq, gq->dirty, num);
+	num = rsw3_get_num_cur_queues(gq);
+	ret = rsw3_gwca_queue_alloc_rx_buf(gq, gq->dirty, num);
 	if (ret < 0)
 		goto err;
-	ret = rswitch_gwca_queue_ext_ts_fill(ndev, gq, gq->dirty, num);
+	ret = rsw3_gwca_queue_ext_ts_fill(ndev, gq, gq->dirty, num);
 	if (ret < 0)
 		goto err;
-	gq->dirty = rswitch_next_queue_index(gq, false, num);
+	gq->dirty = rsw3_next_queue_index(gq, false, num);
 
 	*quota -= limit - boguscnt;
 
 	return boguscnt <= 0;
 
 err:
-	rswitch_gwca_halt(rdev->priv);
+	rsw3_gwca_halt(rdev->priv);
 
 	return 0;
 }
 
-static void rswitch_tx_free(struct net_device *ndev)
+static void rsw3_tx_free(struct net_device *ndev)
 {
-	struct rswitch_device *rdev = netdev_priv(ndev);
-	struct rswitch_gwca_queue *gq = rdev->tx_queue;
-	struct rswitch_ext_desc *desc;
+	struct rsw3_device *rdev = netdev_priv(ndev);
+	struct rsw3_gwca_queue *gq = rdev->tx_queue;
+	struct rsw3_ext_desc *desc;
 	struct sk_buff *skb;
 
-	for (; rswitch_get_num_cur_queues(gq) > 0;
-	     gq->dirty = rswitch_next_queue_index(gq, false, 1)) {
+	for (; rsw3_get_num_cur_queues(gq) > 0;
+	     gq->dirty = rsw3_next_queue_index(gq, false, 1)) {
 		desc = &gq->tx_ring[gq->dirty];
 		if ((desc->desc.die_dt & DT_MASK) != DT_FEMPTY)
 			break;
@@ -891,15 +880,6 @@ static void rswitch_tx_free(struct net_device *ndev)
 		dma_rmb();
 		skb = gq->skbs[gq->dirty];
 		if (skb) {
-			if (skb_shinfo(skb)->tx_flags & SKBTX_HW_TSTAMP) {
-				struct skb_shared_hwtstamps shhwtstamps;
-				struct timespec64 ts;
-
-				rswitch_get_timestamp(rdev->priv, &ts);
-				memset(&shhwtstamps, 0, sizeof(shhwtstamps));
-				shhwtstamps.hwtstamp = timespec64_to_ktime(ts);
-				skb_tstamp_tx(skb, &shhwtstamps);
-			}
 			dma_unmap_single(ndev->dev.parent,
 					 gq->unmap_addrs[gq->dirty],
 					 skb->len, DMA_TO_DEVICE);
@@ -912,11 +892,11 @@ static void rswitch_tx_free(struct net_device *ndev)
 	}
 }
 
-static int rswitch_poll(struct napi_struct *napi, int budget)
+static int rsw3_poll(struct napi_struct *napi, int budget)
 {
 	struct net_device *ndev = napi->dev;
-	struct rswitch_private *priv;
-	struct rswitch_device *rdev;
+	struct rsw3_private *priv;
+	struct rsw3_device *rdev;
 	unsigned long flags;
 	int quota = budget;
 
@@ -924,21 +904,21 @@ static int rswitch_poll(struct napi_struct *napi, int budget)
 	priv = rdev->priv;
 
 retry:
-	rswitch_tx_free(ndev);
+	rsw3_tx_free(ndev);
 
-	if (rswitch_rx(ndev, &quota))
+	if (rsw3_rx(ndev, &quota))
 		goto out;
 	else if (rdev->priv->gwca_halt)
 		goto err;
-	else if (rswitch_is_queue_rxed(rdev->rx_queue))
+	else if (rsw3_is_queue_rxed(rdev->rx_queue))
 		goto retry;
 
 	netif_wake_subqueue(ndev, 0);
 
 	if (napi_complete_done(napi, budget - quota)) {
 		spin_lock_irqsave(&priv->lock, flags);
-		rswitch_enadis_data_irq(priv, rdev->tx_queue->index, true);
-		rswitch_enadis_data_irq(priv, rdev->rx_queue->index, true);
+		rsw3_enadis_data_irq(priv, rdev->tx_queue->index, true);
+		rsw3_enadis_data_irq(priv, rdev->rx_queue->index, true);
 		spin_unlock_irqrestore(&priv->lock, flags);
 	}
 
@@ -951,22 +931,22 @@ static int rswitch_poll(struct napi_struct *napi, int budget)
 	return 0;
 }
 
-static void rswitch_queue_interrupt(struct net_device *ndev)
+static void rsw3_queue_interrupt(struct net_device *ndev)
 {
-	struct rswitch_device *rdev = netdev_priv(ndev);
+	struct rsw3_device *rdev = netdev_priv(ndev);
 
 	if (napi_schedule_prep(&rdev->napi)) {
 		spin_lock(&rdev->priv->lock);
-		rswitch_enadis_data_irq(rdev->priv, rdev->tx_queue->index, false);
-		rswitch_enadis_data_irq(rdev->priv, rdev->rx_queue->index, false);
+		rsw3_enadis_data_irq(rdev->priv, rdev->tx_queue->index, false);
+		rsw3_enadis_data_irq(rdev->priv, rdev->rx_queue->index, false);
 		spin_unlock(&rdev->priv->lock);
 		__napi_schedule(&rdev->napi);
 	}
 }
 
-static irqreturn_t rswitch_data_irq(struct rswitch_private *priv, u32 *dis)
+static irqreturn_t rsw3_data_irq(struct rsw3_private *priv, u32 *dis)
 {
-	struct rswitch_gwca_queue *gq;
+	struct rsw3_gwca_queue *gq;
 	unsigned int i, index, bit;
 
 	for (i = 0; i < priv->gwca.num_queues; i++) {
@@ -976,29 +956,29 @@ static irqreturn_t rswitch_data_irq(struct rswitch_private *priv, u32 *dis)
 		if (!(dis[index] & bit))
 			continue;
 
-		rswitch_ack_data_irq(priv, gq->index);
-		rswitch_queue_interrupt(gq->ndev);
+		rsw3_ack_data_irq(priv, gq->index);
+		rsw3_queue_interrupt(gq->ndev);
 	}
 
 	return IRQ_HANDLED;
 }
 
-static irqreturn_t rswitch_gwca_irq(int irq, void *dev_id)
+static irqreturn_t rsw3_gwca_irq(int irq, void *dev_id)
 {
-	struct rswitch_private *priv = dev_id;
+	struct rsw3_private *priv = dev_id;
 	irqreturn_t ret = IRQ_NONE;
 	u32 dis[RSWITCH3_NUM_IRQ_REGS];
 
-	rswitch_get_data_irq_status(priv, dis);
+	rsw3_get_data_irq_status(priv, dis);
 
-	if (rswitch_is_any_data_irq(priv, dis, true) ||
-	    rswitch_is_any_data_irq(priv, dis, false))
-		ret = rswitch_data_irq(priv, dis);
+	if (rsw3_is_any_data_irq(priv, dis, true) ||
+	    rsw3_is_any_data_irq(priv, dis, false))
+		ret = rsw3_data_irq(priv, dis);
 
 	return ret;
 }
 
-static int rswitch_gwca_request_irqs(struct rswitch_private *priv)
+static int rsw3_gwca_request_irqs(struct rsw3_private *priv)
 {
 	char *resource_name, *irq_name;
 	int i, ret, irq;
@@ -1018,7 +998,7 @@ static int rswitch_gwca_request_irqs(struct rswitch_private *priv)
 		if (!irq_name)
 			return -ENOMEM;
 
-		ret = devm_request_irq(&priv->pdev->dev, irq, rswitch_gwca_irq,
+		ret = devm_request_irq(&priv->pdev->dev, irq, rsw3_gwca_irq,
 				       0, irq_name, priv);
 		if (ret < 0)
 			return ret;
@@ -1028,25 +1008,25 @@ static int rswitch_gwca_request_irqs(struct rswitch_private *priv)
 }
 
 /* Ethernet TSN Agent block (ETHA) and Ethernet MAC IP block (RMAC) */
-static int rswitch_etha_change_mode(struct rswitch_etha *etha,
-				    enum rswitch_etha_mode mode)
+static int rsw3_etha_change_mode(struct rsw3_etha *etha,
+				    enum rsw3_etha_mode mode)
 {
 	int ret;
 
-	if (!rswitch_agent_clock_is_enabled(etha->coma_addr, etha->index))
-		rswitch_agent_clock_ctrl(etha->coma_addr, etha->index, 1);
+	if (!rsw3_agent_clock_is_enabled(etha->coma_addr, etha->index))
+		rsw3_agent_clock_ctrl(etha->coma_addr, etha->index, 1);
 
 	iowrite32(mode, etha->addr + EAMC);
 
-	ret = rswitch_reg_wait(etha->addr, EAMS, EAMS_OPS_MASK, mode);
+	ret = rsw3_reg_wait(etha->addr, EAMS, EAMS_OPS_MASK, mode);
 
 	if (mode == EAMC_OPC_DISABLE)
-		rswitch_agent_clock_ctrl(etha->coma_addr, etha->index, 0);
+		rsw3_agent_clock_ctrl(etha->coma_addr, etha->index, 0);
 
 	return ret;
 }
 
-static void rswitch_etha_read_mac_address(struct rswitch_etha *etha)
+static void rsw3_etha_read_mac_address(struct rsw3_etha *etha)
 {
 	u32 mrmac0 = ioread32(etha->addr + MRMAC0);
 	u32 mrmac1 = ioread32(etha->addr + MRMAC1);
@@ -1060,25 +1040,25 @@ static void rswitch_etha_read_mac_address(struct rswitch_etha *etha)
 	mac[5] = (mrmac1 >>  0) & 0xFF;
 }
 
-static void rswitch_etha_write_mac_address(struct rswitch_etha *etha, const u8 *mac)
+static void rsw3_etha_write_mac_address(struct rsw3_etha *etha, const u8 *mac)
 {
 	iowrite32((mac[0] << 8) | mac[1], etha->addr + MRMAC0);
 	iowrite32((mac[2] << 24) | (mac[3] << 16) | (mac[4] << 8) | mac[5],
 		  etha->addr + MRMAC1);
 }
 
-static int __maybe_unused rswitch_etha_wait_link_verification(struct rswitch_etha *etha)
+static int __maybe_unused rsw3_etha_wait_link_verification(struct rsw3_etha *etha)
 {
 	iowrite32(MLVC_PLV, etha->addr + MLVC);
 
-	return rswitch_reg_wait(etha->addr, MLVC, MLVC_PLV, 0);
+	return rsw3_reg_wait(etha->addr, MLVC, MLVC_PLV, 0);
 }
 
-static void rswitch_rmac_setting(struct rswitch_etha *etha, const u8 *mac)
+static void rsw3_rmac_setting(struct rsw3_etha *etha, const u8 *mac)
 {
 	u32 val;
 
-	rswitch_etha_write_mac_address(etha, mac);
+	rsw3_etha_write_mac_address(etha, mac);
 
 	switch (etha->speed) {
 	case 100:
@@ -1101,168 +1081,161 @@ static void rswitch_rmac_setting(struct rswitch_etha *etha, const u8 *mac)
 	}
 
 	iowrite32(MPIC_PIS_GMII | val, etha->addr + MPIC);
+
+	/* Set MIOC Bit(3)*/
+	if (etha->connect_to_xpcs) {
+		if (etha->index >= 5 && etha->index <= 7) {
+			iowrite32(MIOC_BIT3_SET, etha->addr + MIOC);
+		} else {
+			pr_err("rsw3_rmac_setting: Invalid port %d for XPCS connection\n", etha->index);
+		}
+	}
 }
 
-static void rswitch_etha_enable_mii(struct rswitch_etha *etha)
+static void rsw3_etha_enable_mii(struct rsw3_etha *etha)
 {
-	rswitch_modify(etha->addr, MPIC, MPIC_PSMCS_MASK | MPIC_PSMHT_MASK,
+	rsw3_modify(etha->addr, MPIC, MPIC_PSMCS_MASK | MPIC_PSMHT_MASK,
 		       MPIC_PSMCS(etha->psmcs) | MPIC_PSMHT(0x06));
-	rswitch_modify(etha->addr, MPSM, 0, MPSM_MFF_C45);
+	rsw3_modify(etha->addr, MPSM, 0, MPSM_MMF_C45);
 }
 
-static int rswitch_etha_hw_init(struct rswitch_etha *etha, const u8 *mac)
+static int rsw3_etha_mii_hw_init(struct rsw3_etha *etha)
 {
 	int err;
 
-	err = rswitch_etha_change_mode(etha, EAMC_OPC_DISABLE);
+	if (etha->operated)
+		return 0;
+
+	err = rsw3_etha_change_mode(etha, EAMC_OPC_DISABLE);
 	if (err < 0)
 		return err;
-	err = rswitch_etha_change_mode(etha, EAMC_OPC_CONFIG);
+
+	err = rsw3_etha_change_mode(etha, EAMC_OPC_CONFIG);
 	if (err < 0)
 		return err;
 
-	iowrite32(EAVCC_VEM_SC_TAG, etha->addr + EAVCC);
-	rswitch_rmac_setting(etha, mac);
-	rswitch_etha_enable_mii(etha);
+	rsw3_etha_enable_mii(etha);
 
-	/* Disable function to use vpf environemnt
-	err = rswitch_etha_wait_link_verification(etha);
+	err = rsw3_etha_change_mode(etha, EAMC_OPC_DISABLE);
 	if (err < 0)
 		return err;
-	*/
 
-	err = rswitch_etha_change_mode(etha, EAMC_OPC_DISABLE);
+	err = rsw3_etha_change_mode(etha, EAMC_OPC_OPERATION);
 	if (err < 0)
 		return err;
 
-	return rswitch_etha_change_mode(etha, EAMC_OPC_OPERATION);
+	etha->operated = true;
+
+	return 0;
 }
 
-static int rswitch_etha_set_access(struct rswitch_etha *etha, bool read,
-				   int phyad, int devad, int regad, int data)
+static int rsw3_etha_hw_init(struct rsw3_etha *etha, const u8 *mac)
 {
-	int pop = read ? MDIO_READ_C45 : MDIO_WRITE_C45;
-	u32 val;
-	int ret;
+	int err;
 
-	if (devad == 0xffffffff)
-		return -ENODEV;
+	err = rsw3_etha_change_mode(etha, EAMC_OPC_DISABLE);
+	if (err < 0)
+		return err;
 
-	writel(MMIS1_CLEAR_FLAGS, etha->addr + MMIS1);
+	err = rsw3_etha_change_mode(etha, EAMC_OPC_CONFIG);
+	if (err < 0)
+		return err;
 
-	val = MPSM_PSME | MPSM_MFF_C45;
-	iowrite32((regad << 16) | (devad << 8) | (phyad << 3) | val, etha->addr + MPSM);
+	iowrite32(EAVCC_VEM_SC_TAG, etha->addr + EAVCC);
+	rsw3_rmac_setting(etha, mac);
 
-	ret = rswitch_reg_wait(etha->addr, MMIS1, MMIS1_PAACS, MMIS1_PAACS);
-	if (ret)
-		return ret;
+	err = rsw3_etha_wait_link_verification(etha);
+	if (err < 0)
+		return err;
 
-	rswitch_modify(etha->addr, MMIS1, MMIS1_PAACS, MMIS1_PAACS);
+	err = rsw3_etha_change_mode(etha, EAMC_OPC_DISABLE);
+	if (err < 0)
+		return err;
 
-	if (read) {
-		writel((pop << 13) | (devad << 8) | (phyad << 3) | val, etha->addr + MPSM);
+	return rsw3_etha_change_mode(etha, EAMC_OPC_OPERATION);
+}
 
-		ret = rswitch_reg_wait(etha->addr, MMIS1, MMIS1_PRACS, MMIS1_PRACS);
-		if (ret)
-			return ret;
+static int rsw3_etha_mpsm_op(struct rsw3_etha *etha, bool read,
+				unsigned int mmf, unsigned int pda,
+				unsigned int pra, unsigned int pop,
+				unsigned int prd)
+{
+	u32 val;
+	int ret;
 
-		ret = (ioread32(etha->addr + MPSM) & MPSM_PRD_MASK) >> 16;
+	val = MPSM_PSME |
+	      FIELD_PREP(MPSM_MFF, mmf) |
+	      FIELD_PREP(MPSM_PDA, pda) |
+	      FIELD_PREP(MPSM_PRA, pra) |
+	      FIELD_PREP(MPSM_POP, pop) |
+	      FIELD_PREP(MPSM_PRD, prd);
+	iowrite32(val, etha->addr + MPSM);
 
-		rswitch_modify(etha->addr, MMIS1, MMIS1_PRACS, MMIS1_PRACS);
-	} else {
-		iowrite32((data << 16) | (pop << 13) | (devad << 8) | (phyad << 3) | val,
-			  etha->addr + MPSM);
+	ret = rsw3_reg_wait(etha->addr, MPSM, MPSM_PSME, 0);
+	if (ret)
+		return ret;
 
-		ret = rswitch_reg_wait(etha->addr, MMIS1, MMIS1_PWACS, MMIS1_PWACS);
+	if (read) {
+		val = ioread32(etha->addr + MPSM);
+		ret = FIELD_GET(MPSM_PRD, val);
 	}
 
 	return ret;
 }
 
-static int __maybe_unused rswitch_etha_mii_read_c45(struct mii_bus *bus, int addr,
-						   int devad, int regad)
+static int rsw3_etha_mii_read_c45(struct mii_bus *bus, int addr, int devad,
+				     int regad)
 {
-	struct rswitch_etha *etha = bus->priv;
-
-	return rswitch_etha_set_access(etha, true, addr, devad, regad, 0);
-}
+	struct rsw3_etha *etha = bus->priv;
+	int ret;
 
-static int __maybe_unused rswitch_etha_mii_write_c45(struct mii_bus *bus, int addr,
-						    int devad, int regad, u16 val)
-{
-	struct rswitch_etha *etha = bus->priv;
+	ret = rsw3_etha_mpsm_op(etha, false, MPSM_MMF_C45, addr, devad,
+				   MPSM_POP_ADDRESS, regad);
+	if (ret)
+		return ret;
 
-	return rswitch_etha_set_access(etha, false, addr, devad, regad, val);
+	return rsw3_etha_mpsm_op(etha, true, MPSM_MMF_C45, addr, devad,
+				    MPSM_POP_READ_C45, 0);
 }
 
-static int rswitch_etha_set_access_c22(struct rswitch_etha *etha, bool read,
-				      int phyad, int regad, int data)
+static int rsw3_etha_mii_write_c45(struct mii_bus *bus, int addr, int devad,
+				      int regad, u16 val)
 {
-	int pop = read ? MDIO_READ_C22 : MDIO_WRITE_C22;
+	struct rsw3_etha *etha = bus->priv;
 	int ret;
 
-	rswitch_modify(etha->addr, MPSM, MPSM_POP_MASK, MPSM_POP(pop));
-	rswitch_modify(etha->addr, MPSM, MPSM_PDA_MASK, MPSM_PDA(phyad));
-	rswitch_modify(etha->addr, MPSM, MPSM_PRA_MASK, MPSM_PRA(regad));
-
-	if (!read)
-		rswitch_modify(etha->addr, MPSM, MPSM_PRD_MASK, MPSM_PRD_WRITE(data));
-
-	ret = rswitch_reg_wait(etha->addr, MPSM, MPSM_PSME, 0);
+	ret = rsw3_etha_mpsm_op(etha, false, MPSM_MMF_C45, addr, devad,
+				   MPSM_POP_ADDRESS, regad);
 	if (ret)
 		return ret;
 
-	return read ? MPSM_PRD_READ(ioread32(etha->addr + MPSM)) : 0;
+	return rsw3_etha_mpsm_op(etha, false, MPSM_MMF_C45, addr, devad,
+				    MPSM_POP_WRITE, val);
 }
 
-static int rswitch_etha_mii_read(struct mii_bus *bus, int addr, int regnum)
+static int rsw3_etha_mii_read_c22(struct mii_bus *bus, int phyad, int regad)
 {
-	struct rswitch_etha *etha = bus->priv;
-	struct rswitch_private *priv = container_of(etha,
-			struct rswitch_private, etha[etha->index]);
-	int mode, devad, regad;
-
-	mode = regnum & MII_ADDR_C45;
-	devad = (regnum >> MII_DEVADDR_C45_SHIFT) & 0x1f;
-	regad = regnum & MII_REGADDR_C45_MASK;
+	struct rsw3_etha *etha = bus->priv;
 
-	if (!mode) {
-		if (!priv->vpf_mode)
-			return -EOPNOTSUPP;
-		else
-			return rswitch_etha_set_access_c22(etha, true, addr, regnum, 0);
-	}
-
-	return rswitch_etha_set_access(etha, true, addr, devad, regad, 0);
+	return rsw3_etha_mpsm_op(etha, true, MPSM_MMF_C22, phyad, regad,
+				    MPSM_POP_READ_C22, 0);
 }
 
-static int rswitch_etha_mii_write(struct mii_bus *bus, int addr, int regnum, u16 val)
+static int rsw3_etha_mii_write_c22(struct mii_bus *bus, int phyad,
+				      int regad, u16 val)
 {
-	struct rswitch_etha *etha = bus->priv;
-	struct rswitch_private *priv = container_of(etha,
-			struct rswitch_private, etha[etha->index]);
-	int mode, devad, regad;
+	struct rsw3_etha *etha = bus->priv;
 
-	mode = regnum & MII_ADDR_C45;
-	devad = (regnum >> MII_DEVADDR_C45_SHIFT) & 0x1f;
-	regad = regnum & MII_REGADDR_C45_MASK;
-
-	if (!mode) {
-		if (!priv->vpf_mode)
-			return -EOPNOTSUPP;
-		else
-			return rswitch_etha_set_access_c22(etha, false, addr, regnum, val);
-	}
-
-	return rswitch_etha_set_access(etha, false, addr, devad, regad, val);
+	return rsw3_etha_mpsm_op(etha, false, MPSM_MMF_C22, phyad, regad,
+				    MPSM_POP_WRITE, val);
 }
 
 /* Call of_node_put(port) after done */
-static struct device_node *rswitch_get_port_node(struct rswitch_device *rdev)
+static struct device_node *rsw3_get_port_node(struct rsw3_device *rdev)
 {
 	struct device_node *ports, *port;
 	int err = 0;
-	u32 index;
+	u32 index, etha_mii_idx;
 
 	ports = of_get_child_by_name(rdev->ndev->dev.parent->of_node,
 				     "ethernet-ports");
@@ -1278,6 +1251,13 @@ static struct device_node *rswitch_get_port_node(struct rswitch_device *rdev)
 		if (index == rdev->etha->index) {
 			if (!of_device_is_available(port))
 				port = NULL;
+
+			err = of_property_read_u32(port, "renesas,etha-mii-chan", &etha_mii_idx);
+			if (err < 0)
+				etha_mii_idx = index;
+
+			rdev->etha_mii_idx = etha_mii_idx;
+
 			break;
 		}
 	}
@@ -1288,7 +1268,7 @@ static struct device_node *rswitch_get_port_node(struct rswitch_device *rdev)
 	return port;
 }
 
-static int rswitch_etha_get_params(struct rswitch_device *rdev)
+static int rsw3_etha_get_params(struct rsw3_device *rdev)
 {
 	u32 max_speed;
 	int err;
@@ -1306,6 +1286,8 @@ static int rswitch_etha_get_params(struct rswitch_device *rdev)
 		return 0;
 	}
 
+	rdev->etha->connect_to_xpcs = of_property_read_bool(rdev->np_port, "renesas,connect_to_xpcs");
+
 	/* if no "max-speed" property, let's use default speed */
 	switch (rdev->etha->phy_interface) {
 	case PHY_INTERFACE_MODE_MII:
@@ -1324,7 +1306,7 @@ static int rswitch_etha_get_params(struct rswitch_device *rdev)
 	return 0;
 }
 
-static int rswitch_mii_register(struct rswitch_device *rdev)
+static int rsw3_mii_register(struct rsw3_device *rdev)
 {
 	struct device_node *mdio_np;
 	struct mii_bus *mii_bus;
@@ -1334,11 +1316,13 @@ static int rswitch_mii_register(struct rswitch_device *rdev)
 	if (!mii_bus)
 		return -ENOMEM;
 
-	mii_bus->name = "rswitch_mii";
+	mii_bus->name = "rsw3_mii";
 	sprintf(mii_bus->id, "etha%d", rdev->etha->index);
-	mii_bus->priv = rdev->etha;
-	mii_bus->read = rswitch_etha_mii_read;
-	mii_bus->write = rswitch_etha_mii_write;
+	mii_bus->priv = rdev->etha_mii;
+	mii_bus->read_c45 = rsw3_etha_mii_read_c45;
+	mii_bus->write_c45 = rsw3_etha_mii_write_c45;
+	mii_bus->read = rsw3_etha_mii_read_c22;
+	mii_bus->write = rsw3_etha_mii_write_c22;
 	mii_bus->parent = &rdev->priv->pdev->dev;
 
 	mdio_np = of_get_child_by_name(rdev->np_port, "mdio");
@@ -1356,7 +1340,7 @@ static int rswitch_mii_register(struct rswitch_device *rdev)
 	return err;
 }
 
-static void rswitch_mii_unregister(struct rswitch_device *rdev)
+static void rsw3_mii_unregister(struct rsw3_device *rdev)
 {
 	if (rdev->etha->mii) {
 		mdiobus_unregister(rdev->etha->mii);
@@ -1365,39 +1349,32 @@ static void rswitch_mii_unregister(struct rswitch_device *rdev)
 	}
 }
 
-static void rswitch_adjust_link(struct net_device *ndev)
+static void rsw3_adjust_link(struct net_device *ndev)
 {
-	struct rswitch_device *rdev = netdev_priv(ndev);
+	struct rsw3_device *rdev = netdev_priv(ndev);
 	struct phy_device *phydev = ndev->phydev;
 
 	if (phydev->link != rdev->etha->link) {
 		phy_print_status(phydev);
-		if (!rdev->priv->vpf_mode) {
-			if (phydev->link)
-				phy_power_on(rdev->serdes);
-			else if (rdev->serdes->power_count)
-				phy_power_off(rdev->serdes);
-		}
+		if (phydev->link)
+			phy_power_on(rdev->pcs);
+		else if (rdev->pcs->power_count)
+			phy_power_off(rdev->pcs);
 
 		rdev->etha->link = phydev->link;
 
-		if (!parallel_mode && !rdev->priv->etha_no_runtime_change &&
-		    phydev->speed != rdev->etha->speed) {
+		if (!parallel_mode && phydev->speed != rdev->etha->speed) {
 			rdev->etha->speed = phydev->speed;
 
-			rswitch_etha_hw_init(rdev->etha, rdev->ndev->dev_addr);
-			if (!rdev->priv->vpf_mode)
-				phy_set_speed(rdev->serdes, rdev->etha->speed);
+			rsw3_etha_hw_init(rdev->etha, rdev->ndev->dev_addr);
+			phy_set_speed(rdev->pcs, rdev->etha->speed);
 		}
 	}
 }
 
-static void rswitch_phy_remove_link_mode(struct rswitch_device *rdev,
+static void rsw3_phy_remove_link_mode(struct rsw3_device *rdev,
 					 struct phy_device *phydev)
 {
-	if (!rdev->priv->etha_no_runtime_change)
-		return;
-
 	switch (rdev->etha->speed) {
 	case SPEED_2500:
 		phy_remove_link_mode(phydev, ETHTOOL_LINK_MODE_1000baseT_Full_BIT);
@@ -1418,7 +1395,7 @@ static void rswitch_phy_remove_link_mode(struct rswitch_device *rdev,
 	phy_set_max_speed(phydev, rdev->etha->speed);
 }
 
-static int rswitch_phy_device_init(struct rswitch_device *rdev)
+static int rsw3_phy_device_init(struct rsw3_device *rdev)
 {
 	struct phy_device *phydev;
 	struct device_node *phy;
@@ -1440,7 +1417,7 @@ static int rswitch_phy_device_init(struct rswitch_device *rdev)
 	__set_bit(rdev->etha->phy_interface, phydev->host_interfaces);
 	phydev->mac_managed_pm = true;
 
-	phydev = of_phy_connect(rdev->ndev, phy, rswitch_adjust_link, 0,
+	phydev = of_phy_connect(rdev->ndev, phy, rsw3_adjust_link, 0,
 				rdev->etha->phy_interface);
 	if (!phydev)
 		goto out;
@@ -1450,7 +1427,7 @@ static int rswitch_phy_device_init(struct rswitch_device *rdev)
 	phy_remove_link_mode(phydev, ETHTOOL_LINK_MODE_10baseT_Full_BIT);
 	phy_remove_link_mode(phydev, ETHTOOL_LINK_MODE_100baseT_Half_BIT);
 	phy_remove_link_mode(phydev, ETHTOOL_LINK_MODE_1000baseT_Half_BIT);
-	rswitch_phy_remove_link_mode(rdev, phydev);
+	rsw3_phy_remove_link_mode(rdev, phydev);
 
 	phy_attached_info(phydev);
 
@@ -1461,135 +1438,132 @@ static int rswitch_phy_device_init(struct rswitch_device *rdev)
 	return err;
 }
 
-static void rswitch_phy_device_deinit(struct rswitch_device *rdev)
+static void rsw3_phy_device_deinit(struct rsw3_device *rdev)
 {
 	if (rdev->ndev->phydev)
 		phy_disconnect(rdev->ndev->phydev);
 }
 
-static int rswitch_serdes_set_params(struct rswitch_device *rdev)
+static int rsw3_pcs_set_params(struct rsw3_device *rdev)
 {
 	int err;
 
-	err = phy_set_mode_ext(rdev->serdes, PHY_MODE_ETHERNET,
+	err = phy_set_mode_ext(rdev->pcs, PHY_MODE_ETHERNET,
 			       rdev->etha->phy_interface);
 	if (err < 0)
 		return err;
 
-	return phy_set_speed(rdev->serdes, rdev->etha->speed);
+	return phy_set_speed(rdev->pcs, rdev->etha->speed);
 }
 
-static int rswitch_ether_port_init_one(struct rswitch_device *rdev)
+static int rsw3_ether_port_init_one(struct rsw3_device *rdev)
 {
 	int err;
 
 	if (!parallel_mode) {
-		if (!rdev->etha->operated) {
-			err = rswitch_etha_hw_init(rdev->etha, rdev->ndev->dev_addr);
-			if (err < 0)
-				return err;
-			if (rdev->priv->etha_no_runtime_change)
-				rdev->etha->operated = true;
-
-		err = rswitch_mii_register(rdev);
+		err = rsw3_etha_mii_hw_init(rdev->etha_mii);
 		if (err < 0)
 			return err;
-		if (!rdev->priv->vpf_mode) {
-			err = rswitch_phy_device_init(rdev);
 
-				if (err < 0)
-					goto err_phy_device_init;
+		err = rsw3_etha_hw_init(rdev->etha, rdev->ndev->dev_addr);
+		if (err < 0)
+			return err;
 
-				rdev->serdes = devm_of_phy_get(&rdev->priv->pdev->dev, rdev->np_port, NULL);
-				if (IS_ERR(rdev->serdes)) {
-					err = PTR_ERR(rdev->serdes);
-					goto err_serdes_phy_get;
-				}
+		err = rsw3_mii_register(rdev);
+		if (err < 0)
+			return err;
+		err = rsw3_phy_device_init(rdev);
+		if (err < 0)
+			goto err_phy_device_init;
 
-				err = rswitch_serdes_set_params(rdev);
-				if (err < 0)
-					goto err_serdes_set_params;
-			}
+		rdev->pcs = devm_of_phy_get(&rdev->priv->pdev->dev, rdev->np_port, NULL);
+		if (IS_ERR(rdev->pcs)) {
+			err = -EPROBE_DEFER;
+			goto err_pcs_phy_get;
 		}
+
+		err = rsw3_pcs_set_params(rdev);
+		if (err < 0)
+			goto err_pcs_set_params;
 	}
 
 	return 0;
 
-err_serdes_set_params:
-err_serdes_phy_get:
-	rswitch_phy_device_deinit(rdev);
+err_pcs_set_params:
+err_pcs_phy_get:
+	rsw3_phy_device_deinit(rdev);
 
 err_phy_device_init:
-	rswitch_mii_unregister(rdev);
+	rsw3_mii_unregister(rdev);
 
 	return err;
 }
 
-static void rswitch_ether_port_deinit_one(struct rswitch_device *rdev)
+static void rsw3_ether_port_deinit_one(struct rsw3_device *rdev)
 {
-	rswitch_phy_device_deinit(rdev);
-	rswitch_mii_unregister(rdev);
+	rsw3_phy_device_deinit(rdev);
+	rsw3_mii_unregister(rdev);
 }
 
-static int rswitch_ether_port_init_all(struct rswitch_private *priv)
+static int rsw3_ether_port_init_all(struct rsw3_private *priv)
 {
 	unsigned int i;
 	int err;
 
-	rswitch_for_each_enabled_port(priv, i) {
-		err = rswitch_ether_port_init_one(priv->rdev[i]);
+	rsw3_for_each_enabled_port(priv, i) {
+		if (priv->rdev[i]->disabled)
+			continue;
+		err = rsw3_ether_port_init_one(priv->rdev[i]);
 		if (err)
 			goto err_init_one;
 	}
 
-	if (!priv->vpf_mode) {
-		rswitch_for_each_enabled_port(priv, i) {
-			err = phy_init(priv->rdev[i]->serdes);
-			if (err)
-				goto err_serdes;
-		}
+	rsw3_for_each_enabled_port(priv, i) {
+		if (priv->rdev[i]->disabled)
+			continue;
+		err = phy_init(priv->rdev[i]->pcs);
+		if (err)
+			goto err_pcs;
 	}
 
 	return 0;
 
-err_serdes:
-	rswitch_for_each_enabled_port_continue_reverse(priv, i)
-		phy_exit(priv->rdev[i]->serdes);
+err_pcs:
+	rsw3_for_each_enabled_port_continue_reverse(priv, i)
+		phy_exit(priv->rdev[i]->pcs);
 	i = rswitch3_num_ports;
 
 err_init_one:
-	rswitch_for_each_enabled_port_continue_reverse(priv, i)
-		rswitch_ether_port_deinit_one(priv->rdev[i]);
+	rsw3_for_each_enabled_port_continue_reverse(priv, i)
+		rsw3_ether_port_deinit_one(priv->rdev[i]);
 
 	return err;
 }
 
-static void rswitch_ether_port_deinit_all(struct rswitch_private *priv)
+static void rsw3_ether_port_deinit_all(struct rsw3_private *priv)
 {
 	unsigned int i;
 
 	for (i = 0; i < rswitch3_num_ports; i++) {
-		if (!priv->vpf_mode)
-			phy_exit(priv->rdev[i]->serdes);
+		phy_exit(priv->rdev[i]->pcs);
 
-		rswitch_ether_port_deinit_one(priv->rdev[i]);
+		rsw3_ether_port_deinit_one(priv->rdev[i]);
 	}
 }
 
-static int rswitch_open(struct net_device *ndev)
+static int rsw3_open(struct net_device *ndev)
 {
-	struct rswitch_device *rdev = netdev_priv(ndev);
+	struct rsw3_device *rdev = netdev_priv(ndev);
 	unsigned long flags;
 
-	if (!rdev->priv->vpf_mode)
-		phy_start(ndev->phydev);
+	phy_start(ndev->phydev);
 
 	napi_enable(&rdev->napi);
 	netif_start_queue(ndev);
 
 	spin_lock_irqsave(&rdev->priv->lock, flags);
-	rswitch_enadis_data_irq(rdev->priv, rdev->tx_queue->index, true);
-	rswitch_enadis_data_irq(rdev->priv, rdev->rx_queue->index, true);
+	rsw3_enadis_data_irq(rdev->priv, rdev->tx_queue->index, true);
+	rsw3_enadis_data_irq(rdev->priv, rdev->rx_queue->index, true);
 	spin_unlock_irqrestore(&rdev->priv->lock, flags);
 
 	bitmap_set(rdev->priv->opened_ports, rdev->port, 1);
@@ -1597,29 +1571,28 @@ static int rswitch_open(struct net_device *ndev)
 	return 0;
 };
 
-static int rswitch_stop(struct net_device *ndev)
+static int rsw3_stop(struct net_device *ndev)
 {
-	struct rswitch_device *rdev = netdev_priv(ndev);
+	struct rsw3_device *rdev = netdev_priv(ndev);
 	unsigned long flags;
 
 	netif_tx_stop_all_queues(ndev);
 	bitmap_clear(rdev->priv->opened_ports, rdev->port, 1);
 
 	spin_lock_irqsave(&rdev->priv->lock, flags);
-	rswitch_enadis_data_irq(rdev->priv, rdev->tx_queue->index, false);
-	rswitch_enadis_data_irq(rdev->priv, rdev->rx_queue->index, false);
+	rsw3_enadis_data_irq(rdev->priv, rdev->tx_queue->index, false);
+	rsw3_enadis_data_irq(rdev->priv, rdev->rx_queue->index, false);
 	spin_unlock_irqrestore(&rdev->priv->lock, flags);
 
-	if (!rdev->priv->vpf_mode)
-		phy_stop(ndev->phydev);
+	phy_stop(ndev->phydev);
 	napi_disable(&rdev->napi);
 
 	return 0;
 };
 
-static bool rswitch_ext_desc_set_info1(struct rswitch_device *rdev,
+static bool rsw3_ext_desc_set_info1(struct rsw3_device *rdev,
 				       struct sk_buff *skb,
-				       struct rswitch_ext_desc *desc)
+				       struct rsw3_ext_desc *desc)
 {
 	if (!parallel_mode)
 		desc->info1 = cpu_to_le64(INFO1_DV(BIT(rdev->etha->index)) |
@@ -1641,14 +1614,14 @@ static bool rswitch_ext_desc_set_info1(struct rswitch_device *rdev,
 	return true;
 }
 
-static bool rswitch_ext_desc_set(struct rswitch_device *rdev,
+static bool rsw3_ext_desc_set(struct rsw3_device *rdev,
 				 struct sk_buff *skb,
-				 struct rswitch_ext_desc *desc,
+				 struct rsw3_ext_desc *desc,
 				 dma_addr_t dma_addr, u16 len, u8 die_dt)
 {
-	rswitch_desc_set_dptr(&desc->desc, dma_addr);
+	rsw3_desc_set_dptr(&desc->desc, dma_addr);
 	desc->desc.info_ds = cpu_to_le16(len);
-	if (!rswitch_ext_desc_set_info1(rdev, skb, desc))
+	if (!rsw3_ext_desc_set_info1(rdev, skb, desc))
 		return false;
 
 	dma_wmb();
@@ -1658,7 +1631,7 @@ static bool rswitch_ext_desc_set(struct rswitch_device *rdev,
 	return true;
 }
 
-static u8 rswitch_ext_desc_get_die_dt(unsigned int nr_desc, unsigned int index)
+static u8 rsw3_ext_desc_get_die_dt(unsigned int nr_desc, unsigned int index)
 {
 	if (nr_desc == 1)
 		return DT_FSINGLE | DIE;
@@ -1669,7 +1642,7 @@ static u8 rswitch_ext_desc_get_die_dt(unsigned int nr_desc, unsigned int index)
 	return DT_FMID;
 }
 
-static u16 rswitch_ext_desc_get_len(u8 die_dt, unsigned int orig_len)
+static u16 rsw3_ext_desc_get_len(u8 die_dt, unsigned int orig_len)
 {
 	switch (die_dt & DT_MASK) {
 	case DT_FSINGLE:
@@ -1683,19 +1656,19 @@ static u16 rswitch_ext_desc_get_len(u8 die_dt, unsigned int orig_len)
 	}
 }
 
-static netdev_tx_t rswitch_start_xmit(struct sk_buff *skb, struct net_device *ndev)
+static netdev_tx_t rsw3_start_xmit(struct sk_buff *skb, struct net_device *ndev)
 {
-	struct rswitch_device *rdev = netdev_priv(ndev);
-	struct rswitch_gwca_queue *gq = rdev->tx_queue;
+	struct rsw3_device *rdev = netdev_priv(ndev);
+	struct rsw3_gwca_queue *gq = rdev->tx_queue;
 	dma_addr_t dma_addr, dma_addr_orig;
 	netdev_tx_t ret = NETDEV_TX_OK;
-	struct rswitch_ext_desc *desc;
+	struct rsw3_ext_desc *desc;
 	unsigned int i, nr_desc;
 	u8 die_dt;
 	u16 len;
 
 	nr_desc = (skb->len - 1) / RSWITCH3_DESC_BUF_SIZE + 1;
-	if (rswitch_get_num_cur_queues(gq) >= gq->ring_size - nr_desc) {
+	if (rsw3_get_num_cur_queues(gq) >= gq->ring_size - nr_desc) {
 		netif_stop_subqueue(ndev, 0);
 		return NETDEV_TX_BUSY;
 	}
@@ -1712,18 +1685,18 @@ static netdev_tx_t rswitch_start_xmit(struct sk_buff *skb, struct net_device *nd
 
 	/* DT_FSTART should be set at last. So, this is reverse order. */
 	for (i = nr_desc; i-- > 0; ) {
-		desc = &gq->tx_ring[rswitch_next_queue_index(gq, true, i)];
-		die_dt = rswitch_ext_desc_get_die_dt(nr_desc, i);
+		desc = &gq->tx_ring[rsw3_next_queue_index(gq, true, i)];
+		die_dt = rsw3_ext_desc_get_die_dt(nr_desc, i);
 		dma_addr = dma_addr_orig + i * RSWITCH3_DESC_BUF_SIZE;
-		len = rswitch_ext_desc_get_len(die_dt, skb->len);
-		if (!rswitch_ext_desc_set(rdev, skb, desc, dma_addr, len, die_dt))
+		len = rsw3_ext_desc_get_len(die_dt, skb->len);
+		if (!rsw3_ext_desc_set(rdev, skb, desc, dma_addr, len, die_dt))
 			goto err_unmap;
 	}
 
 	wmb();	/* gq->cur must be incremented after die_dt was set */
 
-	gq->cur = rswitch_next_queue_index(gq, true, nr_desc);
-	rswitch_modify(rdev->addr, GWTRC(gq->index), 0, BIT(gq->index % 32));
+	gq->cur = rsw3_next_queue_index(gq, true, nr_desc);
+	rsw3_modify(rdev->addr, GWTRC(gq->index), 0, BIT(gq->index % 32));
 
 	return ret;
 
@@ -1736,123 +1709,30 @@ static netdev_tx_t rswitch_start_xmit(struct sk_buff *skb, struct net_device *nd
 	return ret;
 }
 
-static struct net_device_stats *rswitch_get_stats(struct net_device *ndev)
+static struct net_device_stats *rsw3_get_stats(struct net_device *ndev)
 {
 	return &ndev->stats;
 }
-static int rswitch_hwstamp_get(struct net_device *ndev, struct ifreq *req)
-{
-	struct rswitch_device *rdev = netdev_priv(ndev);
-	struct rcar_gen4_ptp_private *ptp_priv;
-	struct hwtstamp_config config;
-
-	ptp_priv = rdev->priv->ptp_priv;
-
-	config.flags = 0;
-	config.tx_type = ptp_priv->tstamp_tx_ctrl ? HWTSTAMP_TX_ON :
-						    HWTSTAMP_TX_OFF;
-	switch (ptp_priv->tstamp_rx_ctrl & RCAR_GEN4_RXTSTAMP_TYPE) {
-	case RCAR_GEN4_RXTSTAMP_TYPE_V2_L2_EVENT:
-		config.rx_filter = HWTSTAMP_FILTER_PTP_V2_L2_EVENT;
-		break;
-	case RCAR_GEN4_RXTSTAMP_TYPE_ALL:
-		config.rx_filter = HWTSTAMP_FILTER_ALL;
-		break;
-	default:
-		config.rx_filter = HWTSTAMP_FILTER_NONE;
-		break;
-	}
-
-	return copy_to_user(req->ifr_data, &config, sizeof(config)) ? -EFAULT : 0;
-}
-
-static int rswitch_hwstamp_set(struct net_device *ndev, struct ifreq *req)
-{
-	struct rswitch_device *rdev = netdev_priv(ndev);
-	u32 tstamp_rx_ctrl = RCAR_GEN4_RXTSTAMP_ENABLED;
-	struct hwtstamp_config config;
-	u32 tstamp_tx_ctrl;
-
-	if (copy_from_user(&config, req->ifr_data, sizeof(config)))
-		return -EFAULT;
-
-	if (config.flags)
-		return -EINVAL;
-
-	switch (config.tx_type) {
-	case HWTSTAMP_TX_OFF:
-		tstamp_tx_ctrl = 0;
-		break;
-	case HWTSTAMP_TX_ON:
-		tstamp_tx_ctrl = RCAR_GEN4_TXTSTAMP_ENABLED;
-		break;
-	default:
-		return -ERANGE;
-	}
-
-	switch (config.rx_filter) {
-	case HWTSTAMP_FILTER_NONE:
-		tstamp_rx_ctrl = 0;
-		break;
-	case HWTSTAMP_FILTER_PTP_V2_L2_EVENT:
-		tstamp_rx_ctrl |= RCAR_GEN4_RXTSTAMP_TYPE_V2_L2_EVENT;
-		break;
-	default:
-		config.rx_filter = HWTSTAMP_FILTER_ALL;
-		tstamp_rx_ctrl |= RCAR_GEN4_RXTSTAMP_TYPE_ALL;
-		break;
-	}
-
-	rdev->priv->ptp_priv->tstamp_tx_ctrl = tstamp_tx_ctrl;
-	rdev->priv->ptp_priv->tstamp_rx_ctrl = tstamp_rx_ctrl;
-
-	return copy_to_user(req->ifr_data, &config, sizeof(config)) ? -EFAULT : 0;
-}
 
-static int rswitch_eth_ioctl(struct net_device *ndev, struct ifreq *req, int cmd)
+static int rsw3_eth_ioctl(struct net_device *ndev, struct ifreq *req, int cmd)
 {
 	if (!netif_running(ndev))
 		return -EINVAL;
 
-	switch (cmd) {
-	case SIOCGHWTSTAMP:
-		return rswitch_hwstamp_get(ndev, req);
-	case SIOCSHWTSTAMP:
-		return rswitch_hwstamp_set(ndev, req);
-	default:
-		return phy_mii_ioctl(ndev->phydev, req, cmd);
-	}
+	return phy_mii_ioctl(ndev->phydev, req, cmd);
 }
 
-static const struct net_device_ops rswitch_netdev_ops = {
-	.ndo_open = rswitch_open,
-	.ndo_stop = rswitch_stop,
-	.ndo_start_xmit = rswitch_start_xmit,
-	.ndo_get_stats = rswitch_get_stats,
-	.ndo_eth_ioctl = rswitch_eth_ioctl,
+static const struct net_device_ops rsw3_netdev_ops = {
+	.ndo_open = rsw3_open,
+	.ndo_stop = rsw3_stop,
+	.ndo_start_xmit = rsw3_start_xmit,
+	.ndo_get_stats = rsw3_get_stats,
+	.ndo_eth_ioctl = rsw3_eth_ioctl,
 	.ndo_validate_addr = eth_validate_addr,
 	.ndo_set_mac_address = eth_mac_addr,
 };
 
-static int rswitch_get_ts_info(struct net_device *ndev, struct ethtool_ts_info *info)
-{
-	struct rswitch_device *rdev = netdev_priv(ndev);
-
-	info->phc_index = ptp_clock_index(rdev->priv->ptp_priv->clock);
-	info->so_timestamping = SOF_TIMESTAMPING_TX_SOFTWARE |
-				SOF_TIMESTAMPING_RX_SOFTWARE |
-				SOF_TIMESTAMPING_SOFTWARE |
-				SOF_TIMESTAMPING_TX_HARDWARE |
-				SOF_TIMESTAMPING_RX_HARDWARE |
-				SOF_TIMESTAMPING_RAW_HARDWARE;
-	info->tx_types = BIT(HWTSTAMP_TX_OFF) | BIT(HWTSTAMP_TX_ON);
-	info->rx_filters = BIT(HWTSTAMP_FILTER_NONE) | BIT(HWTSTAMP_FILTER_ALL);
-
-	return 0;
-}
-
-static const struct ethtool_ops rswitch_ethtool_ops = {
-	.get_ts_info = rswitch_get_ts_info,
+static const struct ethtool_ops rsw3_ethtool_ops = {
 	.get_link_ksettings = phy_ethtool_get_link_ksettings,
 	.set_link_ksettings = phy_ethtool_set_link_ksettings,
 };
@@ -1863,33 +1743,39 @@ static const struct of_device_id renesas_eth_sw_of_table[] = {
 };
 MODULE_DEVICE_TABLE(of, renesas_eth_sw_of_table);
 
-static void rswitch_etha_init(struct rswitch_private *priv, unsigned int index)
+static void rsw3_etha_init(struct rsw3_private *priv, unsigned int index)
 {
-	struct rswitch_etha *etha = &priv->etha[index];
+	struct rsw3_etha *etha = &priv->etha[index];
+	struct rsw3_etha *etha_mii = &priv->etha_mii[index];
 
 	memset(etha, 0, sizeof(*etha));
 	etha->index = index;
 	etha->addr = priv->addr + RSWITCH3_ETHA_OFFSET + index * RSWITCH3_ETHA_SIZE;
 	etha->coma_addr = priv->addr;
 
+	memset(etha_mii, 0, sizeof(*etha));
+	etha_mii->index = index;
+	etha_mii->addr = priv->addr + RSWITCH3_ETHA_OFFSET + index * RSWITCH3_ETHA_SIZE;
+	etha_mii->coma_addr = priv->addr;
+
 	/* MPIC.PSMCS = (clk [MHz] / (MDC frequency [MHz] * 2) - 1.
 	 * Calculating PSMCS value as MDC frequency = 2.5MHz. So, multiply
 	 * both the numerator and the denominator by 10.
 	 */
-	etha->psmcs = clk_get_rate(priv->clk) / 100000 / (25 * 2) - 1;
+	etha_mii->psmcs = clk_get_rate(priv->clk) / 100000 / (25 * 2) - 1;
 }
 
-static int rswitch_device_alloc(struct rswitch_private *priv, unsigned int index)
+static int rsw3_device_alloc(struct rsw3_private *priv, unsigned int index)
 {
 	struct platform_device *pdev = priv->pdev;
-	struct rswitch_device *rdev;
+	struct rsw3_device *rdev;
 	struct net_device *ndev;
 	int err;
 
 	if (index >= rswitch3_num_ports)
 		return -EINVAL;
 
-	ndev = alloc_etherdev_mqs(sizeof(struct rswitch_device), 1, 1);
+	ndev = alloc_etherdev_mqs(sizeof(struct rsw3_device), 1, 1);
 	if (!ndev)
 		return -ENOMEM;
 
@@ -1902,21 +1788,22 @@ static int rswitch_device_alloc(struct rswitch_private *priv, unsigned int index
 	priv->rdev[index] = rdev;
 	rdev->port = index;
 	rdev->etha = &priv->etha[index];
+	rdev->etha_mii = &priv->etha_mii[index];
 	rdev->addr = priv->addr;
 
 	ndev->base_addr = (unsigned long)rdev->addr;
 	snprintf(ndev->name, IFNAMSIZ, "tsn%d", index);
-	ndev->netdev_ops = &rswitch_netdev_ops;
-	ndev->ethtool_ops = &rswitch_ethtool_ops;
+	ndev->netdev_ops = &rsw3_netdev_ops;
 	ndev->max_mtu = RSWITCH3_MAX_MTU;
 	ndev->min_mtu = ETH_MIN_MTU;
 
-	netif_napi_add(ndev, &rdev->napi, rswitch_poll);
+	netif_napi_add(ndev, &rdev->napi, rsw3_poll);
 
-	rdev->np_port = rswitch_get_port_node(rdev);
+	rdev->np_port = rsw3_get_port_node(rdev);
 	rdev->disabled = !rdev->np_port;
 	err = of_get_ethdev_address(rdev->np_port, ndev);
 	of_node_put(rdev->np_port);
+
 	if (err) {
 		if (is_valid_ether_addr(rdev->etha->mac_addr))
 			eth_hw_addr_set(ndev, rdev->etha->mac_addr);
@@ -1924,25 +1811,27 @@ static int rswitch_device_alloc(struct rswitch_private *priv, unsigned int index
 			eth_hw_addr_random(ndev);
 	}
 
-	err = rswitch_etha_get_params(rdev);
+	rdev->etha_mii = &priv->etha_mii[rdev->etha_mii_idx];
+
+	err = rsw3_etha_get_params(rdev);
 	if (err < 0)
 		goto out_get_params;
 
 	if (rdev->priv->gwca.speed < rdev->etha->speed)
 		rdev->priv->gwca.speed = rdev->etha->speed;
 
-	err = rswitch_rxdmac_alloc(ndev);
+	err = rsw3_rxdmac_alloc(ndev);
 	if (err < 0)
 		goto out_rxdmac;
 
-	err = rswitch_txdmac_alloc(ndev);
+	err = rsw3_txdmac_alloc(ndev);
 	if (err < 0)
 		goto out_txdmac;
 
 	return 0;
 
 out_txdmac:
-	rswitch_rxdmac_free(ndev);
+	rsw3_rxdmac_free(ndev);
 
 out_rxdmac:
 out_get_params:
@@ -1952,129 +1841,115 @@ static int rswitch_device_alloc(struct rswitch_private *priv, unsigned int index
 	return err;
 }
 
-static void rswitch_device_free(struct rswitch_private *priv, unsigned int index)
+static void rsw3_device_free(struct rsw3_private *priv, unsigned int index)
 {
-	struct rswitch_device *rdev = priv->rdev[index];
+	struct rsw3_device *rdev = priv->rdev[index];
 	struct net_device *ndev = rdev->ndev;
 
-	rswitch_txdmac_free(ndev);
-	rswitch_rxdmac_free(ndev);
+	if (!priv->rdev[index]->disabled) {
+		rsw3_txdmac_free(ndev);
+		rsw3_rxdmac_free(ndev);
+	}
+
 	netif_napi_del(&rdev->napi);
 	free_netdev(ndev);
 }
 
-static int rswitch_init(struct rswitch_private *priv)
+static int rsw3_init(struct rsw3_private *priv)
 {
 	unsigned int i;
 	int err;
 
 	for (i = 0; i < rswitch3_num_ports; i++)
-		rswitch_etha_init(priv, i);
+		rsw3_etha_init(priv, i);
 
 	if (!parallel_mode)
-		rswitch_clock_enable(priv);
+		rsw3_clock_enable(priv);
 	for (i = 0; i < rswitch3_num_ports; i++)
-		rswitch_etha_read_mac_address(&priv->etha[i]);
+		rsw3_etha_read_mac_address(&priv->etha[i]);
 
-	rswitch_reset(priv);
+	rsw3_reset(priv);
 
 	if (!parallel_mode)
-		rswitch_clock_enable(priv);
+		rsw3_clock_enable(priv);
 
-	rswitch_top_init(priv);
+	rsw3_top_init(priv);
 	if (!parallel_mode) {
-		err = rswitch_bpool_config(priv);
+		err = rsw3_bpool_config(priv);
 		if (err < 0)
 			return err;
 
-		rswitch_coma_init(priv);
+		rsw3_coma_init(priv);
 
-		err = rswitch_gwca_linkfix_alloc(priv);
+		err = rsw3_gwca_linkfix_alloc(priv);
 		if (err < 0)
 			return -ENOMEM;
 	}
 
 	for (i = 0; i < rswitch3_num_ports; i++) {
-		err = rswitch_device_alloc(priv, i);
+		err = rsw3_device_alloc(priv, i);
 		if (err < 0) {
 			for (; i-- > 0; )
-				rswitch_device_free(priv, i);
+				rsw3_device_free(priv, i);
 		}
 	}
 
 	if (!parallel_mode)
-		rswitch_fwd_init(priv);
-
-	err = rcar_gen4_ptp_register(priv->ptp_priv, RCAR_GEN4_PTP_REG_LAYOUT,
-				     RCAR_GEN4_PTP_CLOCK_S4);
-	if (err < 0)
-		goto err_ptp_register;
+		rsw3_fwd_init(priv);
 
-	err = rswitch_gwca_request_irqs(priv);
+	err = rsw3_gwca_request_irqs(priv);
 	if (err < 0)
 		goto err_gwca_request_irq;
 
-
-	err = rswitch_gwca_hw_init(priv);
+	err = rsw3_gwca_hw_init(priv);
 	if (err < 0)
 		goto err_gwca_hw_init;
 
-	err = rswitch_ether_port_init_all(priv);
+	err = rsw3_ether_port_init_all(priv);
 	if (err)
 		goto err_ether_port_init_all;
 
-	rswitch_for_each_enabled_port(priv, i) {
+	rsw3_for_each_enabled_port(priv, i) {
 		err = register_netdev(priv->rdev[i]->ndev);
 		if (err) {
-			rswitch_for_each_enabled_port_continue_reverse(priv, i)
+			rsw3_for_each_enabled_port_continue_reverse(priv, i)
 				unregister_netdev(priv->rdev[i]->ndev);
 			goto err_register_netdev;
 		}
 	}
 
-	rswitch_for_each_enabled_port(priv, i)
+	rsw3_for_each_enabled_port(priv, i)
 		netdev_info(priv->rdev[i]->ndev, "MAC address %pM\n",
 			    priv->rdev[i]->ndev->dev_addr);
 
 	return 0;
 
 err_register_netdev:
-	rswitch_ether_port_deinit_all(priv);
+	rsw3_ether_port_deinit_all(priv);
 
 err_ether_port_init_all:
-	rswitch_gwca_hw_deinit(priv);
+	rsw3_gwca_hw_deinit(priv);
 
 err_gwca_hw_init:
 err_gwca_request_irq:
-	rcar_gen4_ptp_unregister(priv->ptp_priv);
-
-err_ptp_register:
 	for (i = 0; i < rswitch3_num_ports; i++)
-		rswitch_device_free(priv, i);
+		rsw3_device_free(priv, i);
 	return err;
 }
 
 static int renesas_eth_sw_probe(struct platform_device *pdev)
 {
-	struct rswitch_private *priv;
-	struct resource *res, *res_ptp;
+	struct rsw3_private *priv;
+	struct resource *res;
 	int ret;
 
 	res = platform_get_resource_byname(pdev, IORESOURCE_MEM, "secure_base");
-	res_ptp = platform_get_resource_byname(pdev, IORESOURCE_MEM, "gptp");
-	if (!res || !res_ptp) {
-		dev_err(&pdev->dev, "invalid resource\n");
+	if (!res)
 		return -EINVAL;
-	}
 
 	priv = devm_kzalloc(&pdev->dev, sizeof(*priv), GFP_KERNEL);
 	if (!priv)
 		return -ENOMEM;
-	spin_lock_init(&priv->lock);
-
-	priv->clk = devm_clk_get(&pdev->dev, NULL);
-	if (IS_ERR(priv->clk))
-		return PTR_ERR(priv->clk);
 
 	if (!parallel_mode)
 		parallel_mode = of_property_read_bool(pdev->dev.of_node, "parallel_mode");
@@ -2082,41 +1957,12 @@ static int renesas_eth_sw_probe(struct platform_device *pdev)
 	if (parallel_mode)
 		rswitch3_num_ports = 1;
 
-	if (of_find_property(pdev->dev.of_node, "vpf_mode", NULL))
-		priv->vpf_mode = true;
-	else
-		priv->vpf_mode = false;
-
-	if (!parallel_mode) {
-		priv->rsw_clk = devm_clk_get(&pdev->dev, "rsw2");
-		if (IS_ERR(priv->rsw_clk)) {
-			dev_err(&pdev->dev, "Failed to get rsw2 clock: %ld\n",
-				PTR_ERR(priv->rsw_clk));
-			return -PTR_ERR(priv->rsw_clk);
-		}
-
-		priv->phy_clk = devm_clk_get(&pdev->dev, "eth-phy");
-		if (IS_ERR(priv->phy_clk)) {
-			dev_err(&pdev->dev, "Failed to get eth-phy clock: %ld\n",
-				PTR_ERR(priv->phy_clk));
-			return -PTR_ERR(priv->phy_clk);
-		}
-	}
-
 	platform_set_drvdata(pdev, priv);
 	priv->pdev = pdev;
+
 	priv->addr = devm_ioremap_resource(&pdev->dev, res);
 	if (IS_ERR(priv->addr))
 		return PTR_ERR(priv->addr);
-	priv->ptp_priv = rcar_gen4_ptp_alloc(pdev);
-	if (!priv->ptp_priv)
-		return -ENOMEM;
-
-	priv->ptp_priv->addr = devm_ioremap_resource(&pdev->dev, res_ptp);
-	if (IS_ERR(priv->ptp_priv->addr))
-		return PTR_ERR(priv->ptp_priv->addr);
-
-	priv->ptp_priv->parallel_mode = parallel_mode;
 
 	ret = dma_set_mask_and_coherent(&pdev->dev, DMA_BIT_MASK(40));
 	if (ret < 0) {
@@ -2138,7 +1984,7 @@ static int renesas_eth_sw_probe(struct platform_device *pdev)
 		pm_runtime_get_sync(&pdev->dev);
 	}
 
-	ret = rswitch_init(priv);
+	ret = rsw3_init(priv);
 	if (ret < 0) {
 		if (!parallel_mode) {
 			pm_runtime_put(&pdev->dev);
@@ -2152,35 +1998,33 @@ static int renesas_eth_sw_probe(struct platform_device *pdev)
 	return ret;
 }
 
-static void rswitch_deinit(struct rswitch_private *priv)
+static void rsw3_deinit(struct rsw3_private *priv)
 {
 	unsigned int i;
 
-	rswitch_gwca_hw_deinit(priv);
-	rcar_gen4_ptp_unregister(priv->ptp_priv);
+	rsw3_gwca_hw_deinit(priv);
 
-	rswitch_for_each_enabled_port(priv, i) {
-		struct rswitch_device *rdev = priv->rdev[i];
+	rsw3_for_each_enabled_port(priv, i) {
+		struct rsw3_device *rdev = priv->rdev[i];
 
 		unregister_netdev(rdev->ndev);
-		rswitch_ether_port_deinit_one(rdev);
-		if (!priv->vpf_mode)
-			phy_exit(priv->rdev[i]->serdes);
+		rsw3_ether_port_deinit_one(rdev);
+		phy_exit(priv->rdev[i]->pcs);
 	}
 
 	for (i = 0; i < rswitch3_num_ports; i++)
-		rswitch_device_free(priv, i);
+		rsw3_device_free(priv, i);
 
-	rswitch_gwca_linkfix_free(priv);
+	rsw3_gwca_linkfix_free(priv);
 
-	rswitch_clock_disable(priv);
+	rsw3_clock_disable(priv);
 }
 
 static void renesas_eth_sw_remove(struct platform_device *pdev)
 {
-	struct rswitch_private *priv = platform_get_drvdata(pdev);
+	struct rsw3_private *priv = platform_get_drvdata(pdev);
 
-	rswitch_deinit(priv);
+	rsw3_deinit(priv);
 
 	pm_runtime_put(&pdev->dev);
 	pm_runtime_disable(&pdev->dev);
@@ -2190,21 +2034,19 @@ static void renesas_eth_sw_remove(struct platform_device *pdev)
 
 static int renesas_eth_sw_suspend(struct device *dev)
 {
-	struct rswitch_private *priv = dev_get_drvdata(dev);
+	struct rsw3_private *priv = dev_get_drvdata(dev);
 	struct net_device *ndev;
 	unsigned int i;
 
-	rswitch_for_each_enabled_port(priv, i) {
+	rsw3_for_each_enabled_port(priv, i) {
 		ndev = priv->rdev[i]->ndev;
 		if (netif_running(ndev)) {
 			netif_device_detach(ndev);
-			rswitch_stop(ndev);
+			rsw3_stop(ndev);
 		}
 
-		if (!priv->vpf_mode) {
-			if (priv->rdev[i]->serdes->init_count)
-				phy_exit(priv->rdev[i]->serdes);
-		}
+		if (priv->rdev[i]->pcs->init_count)
+			phy_exit(priv->rdev[i]->pcs);
 	}
 
 	return 0;
@@ -2212,17 +2054,16 @@ static int renesas_eth_sw_suspend(struct device *dev)
 
 static int renesas_eth_sw_resume(struct device *dev)
 {
-	struct rswitch_private *priv = dev_get_drvdata(dev);
+	struct rsw3_private *priv = dev_get_drvdata(dev);
 	struct net_device *ndev;
 	unsigned int i;
 
-	rswitch_for_each_enabled_port(priv, i) {
-		if (!priv->vpf_mode)
-			phy_init(priv->rdev[i]->serdes);
+	rsw3_for_each_enabled_port(priv, i) {
+		phy_init(priv->rdev[i]->pcs);
 
 		ndev = priv->rdev[i]->ndev;
 		if (netif_running(ndev)) {
-			rswitch_open(ndev);
+			rsw3_open(ndev);
 			netif_device_attach(ndev);
 		}
 	}
@@ -2243,5 +2084,5 @@ static struct platform_driver renesas_eth_sw_driver_platform = {
 	}
 };
 module_platform_driver(renesas_eth_sw_driver_platform);
-MODULE_DESCRIPTION("Renesas Ethernet Switch device driver");
+MODULE_DESCRIPTION("Renesas Ethernet Switch3 device driver");
 MODULE_LICENSE("GPL");
diff --git a/drivers/net/ethernet/renesas/rswitch3.h b/drivers/net/ethernet/renesas/rswitch3.h
index 6a9d58702242..8b2f54366e5c 100644
--- a/drivers/net/ethernet/renesas/rswitch3.h
+++ b/drivers/net/ethernet/renesas/rswitch3.h
@@ -1,16 +1,16 @@
-/* SPDX-License-Identifier: GPL-2.0 */
-/* Renesas Ethernet Switch3 Device Driver
- * This driver is based on the Ethernet Switch2 driver
- * from Linux Kernel v6.11-rc2
+/* SPDX-License-Identifier: GPL-2.0-only */
+/*
+ * Header for Renesas Ethernet Switch3.
+ *
+ * Copyright (C) 2025 Renesas Electronics Corporation
  *
- * Copyright (C) 2024 Renesas Electronics Corporation
+ * Based on the Renesas Ethernet Switch2 on R-Car S4-8
  */
 
 #ifndef __RSWITCH3_H__
 #define __RSWITCH3_H__
 
 #include <linux/platform_device.h>
-#include "rcar_gen4_ptp.h"
 
 #define RSWITCH3_MAX_NUM_QUEUES		128
 
@@ -23,13 +23,13 @@
 #define RSWITCH3_GWCA_IDX_TO_HW_NUM(i)	((i) + RSWITCH3_NUM_PORTS)
 #define RSWITCH3_HW_NUM_TO_GWCA_IDX(i)	((i) - RSWITCH3_NUM_PORTS)
 
-#define rswitch_for_each_enabled_port(priv, i)		\
+#define rsw3_for_each_enabled_port(priv, i)		\
 	for (i = 0; i < RSWITCH3_NUM_PORTS; i++)		\
 		if (priv->rdev[i]->disabled)		\
 			continue;			\
 		else
 
-#define rswitch_for_each_enabled_port_continue_reverse(priv, i)	\
+#define rsw3_for_each_enabled_port_continue_reverse(priv, i)	\
 	for (; i-- > 0; )					\
 		if (priv->rdev[i]->disabled)			\
 			continue;				\
@@ -48,31 +48,31 @@
 #define RSWITCH3_MAP_BUF_SIZE	(RSWITCH3_BUF_SIZE - RSWITCH3_HEADROOM)
 #define RSWITCH3_MAX_CTAG_PCP	7
 
-#define RSWITCH3_TIMEOUT_US		10000000
+#define RSWITCH3_TIMEOUT_US	10000000
 
-#define RSWITCH3_TOP_OFFSET		0x1b000
+#define RSWITCH3_TOP_OFFSET	0x1b000
 #define RSWITCH3_COMA_OFFSET	0x1c000
 #define RSWITCH3_ETHA_OFFSET	0x1d000 /* with RMAC */
-#define RSWITCH3_ETHA_SIZE		0x02000 /* with RMAC */
+#define RSWITCH3_ETHA_SIZE	0x02000 /* with RMAC */
 #define RSWITCH3_GWCA0_OFFSET	0x37000
 #define RSWITCH3_GWCA1_OFFSET	0x39000
 
 
 /* TODO: hardcoded ETHA/GWCA1 settings for now */
-#define GWCA_IRQ_RESOURCE_NAME		"gwca1_gwdis%d"
-#define GWCA_IRQ_NAME				"rswitch3: gwca1_gwdis%d"
-#define GWCA_NUM_IRQS				1
-#define GWCA_INDEX					1
-#define AGENT_INDEX_GWCA			14
-#define GWCA_IPV_NUM				0
-#define GWRO						RSWITCH3_GWCA1_OFFSET
+#define GWCA_IRQ_RESOURCE_NAME	"gwca1_gwdis%d"
+#define GWCA_IRQ_NAME		"rswitch3: gwca1_gwdis%d"
+#define GWCA_NUM_IRQS		1
+#define GWCA_INDEX		1
+#define AGENT_INDEX_GWCA	14
+#define GWCA_IPV_NUM		0
+#define GWRO			RSWITCH3_GWCA1_OFFSET
 
 #define FWRO	0
 #define TPRO	RSWITCH3_TOP_OFFSET
 #define CARO	RSWITCH3_COMA_OFFSET
 #define TARO	0
 #define RMRO	0x1000
-enum rswitch_reg {
+enum rsw3_reg {
 	FWGC			= FWRO + 0x0000,
 	FWTTC0			= FWRO + 0x0010,
 	FWTTC1			= FWRO + 0x0014,
@@ -89,7 +89,7 @@ enum rswitch_reg {
 	FWEMPL23URC		= FWRO + 0x004c,
 	FWSDMPTC		= FWRO + 0x0050,
 	FWSDMPVC		= FWRO + 0x0054,
-	FWSDMPL23URC	= FWRO + 0x0058,
+	FWSDMPL23URC		= FWRO + 0x0058,
 	FWSMPTC			= FWRO + 0x0060,
 	FWSMPVC			= FWRO + 0x0064,
 	FWSMPL23URC		= FWRO + 0x0068,
@@ -485,7 +485,7 @@ enum rswitch_reg {
 	TEIM0			= TPRO + 0x0950,
 
 	RIPV			= CARO + 0x0000,
-	RRC				= CARO + 0x0004,
+	RRC			= CARO + 0x0004,
 	RCEC			= CARO + 0x0008,
 	RCDC			= CARO + 0x000c,
 	RSSIS			= CARO + 0x0010,
@@ -722,6 +722,7 @@ enum rswitch_reg {
 	GWVCC			= GWRO + 0x0130,
 	GWVTC			= GWRO + 0x0134,
 	GWTTFC			= GWRO + 0x0138,
+	GWCKSC			= GWRO + 0x013c,
 	GWTDCAC00		= GWRO + 0x0140,
 	GWTDCAC10		= GWRO + 0x0144,
 	GWTSDCC0		= GWRO + 0x0160,
@@ -793,7 +794,7 @@ enum rswitch_reg {
 };
 
 /* ETHA/RMAC */
-enum rswitch_etha_mode {
+enum rsw3_etha_mode {
 	EAMC_OPC_RESET,
 	EAMC_OPC_DISABLE,
 	EAMC_OPC_CONFIG,
@@ -809,36 +810,28 @@ enum rswitch_etha_mode {
 #define MPIC_PIS_XGMII		0x04
 #define MPIC_LSC_SHIFT		3
 #define MPIC_LSC_100M		(1 << MPIC_LSC_SHIFT)
-#define MPIC_LSC_1G			(2 << MPIC_LSC_SHIFT)
+#define MPIC_LSC_1G		(2 << MPIC_LSC_SHIFT)
 #define MPIC_LSC_2_5G		(3 << MPIC_LSC_SHIFT)
-#define MPIC_LSC_5G			(4 << MPIC_LSC_SHIFT)
+#define MPIC_LSC_5G		(4 << MPIC_LSC_SHIFT)
 #define MPIC_LSC_10G		(5 << MPIC_LSC_SHIFT)
 
-#define MDIO_READ_C45		0x03
-#define MDIO_WRITE_C45		0x01
-#define MDIO_READ_C22		0x02
-#define MDIO_WRITE_C22		0x01
-
-#define MPSM_PSME			BIT(0)
-#define MPSM_MFF_C45		BIT(2)
-#define MPSM_PDA_SHIFT		3
-#define MPSM_PDA_MASK		GENMASK(7, MPSM_PDA_SHIFT)
-#define MPSM_PDA(val)		((val) << MPSM_PDA_SHIFT)
-#define MPSM_PRA_SHIFT		8
-#define MPSM_PRA_MASK		GENMASK(12, MPSM_PRA_SHIFT)
-#define MPSM_PRA(val)		((val) << MPSM_PRA_SHIFT)
-#define MPSM_POP_SHIFT		13
-#define MPSM_POP_MASK		GENMASK(14, MPSM_POP_SHIFT)
-#define MPSM_POP(val)		((val) << MPSM_POP_SHIFT)
-#define MPSM_PRD_SHIFT		16
-#define MPSM_PRD_MASK		GENMASK(31, MPSM_PRD_SHIFT)
-#define MPSM_PRD_WRITE(val)	((val) << MPSM_PRD_SHIFT)
-#define MPSM_PRD_READ(val)	((val) & MPSM_PRD_MASK >> MPSM_PRD_SHIFT)
+#define MPSM_PSME		BIT(0)
+#define MPSM_MFF		BIT(2)
+#define MPSM_MMF_C22		0
+#define MPSM_MMF_C45		1
+#define MPSM_PDA		GENMASK(7, 3)
+#define MPSM_PRA		GENMASK(12, 8)
+#define MPSM_POP		GENMASK(14, 13)
+#define MPSM_POP_ADDRESS	0
+#define MPSM_POP_WRITE		1
+#define MPSM_POP_READ_C22	2
+#define MPSM_POP_READ_C45	3
+#define MPSM_PRD		GENMASK(31, 16)
 
 /* Completion flags */
-#define MMIS1_PAACS			BIT(2) /* Address */
-#define MMIS1_PWACS			BIT(1) /* Write */
-#define MMIS1_PRACS			BIT(0) /* Read */
+#define MMIS1_PAACS		BIT(2) /* Address */
+#define MMIS1_PWACS		BIT(1) /* Write */
+#define MMIS1_PRACS		BIT(0) /* Read */
 #define MMIS1_CLEAR_FLAGS 	0xf
 
 #define MPIC_PSMCS_SHIFT	16
@@ -851,8 +844,11 @@ enum rswitch_etha_mode {
 
 #define MLVC_PLV		BIT(16)
 
+/* MIOC Register define */
+#define MIOC_BIT3_SET		BIT(3)
+
 /* GWCA */
-enum rswitch_gwca_mode {
+enum rsw3_gwca_mode {
 	GWMC_OPC_RESET,
 	GWMC_OPC_DISABLE,
 	GWMC_OPC_CONFIG,
@@ -862,72 +858,73 @@ enum rswitch_gwca_mode {
 #define GWMS_OPS_MASK		GWMC_OPC_OPERATION
 
 #define GWMTIRM_MTIOG		BIT(0)
-#define GWMTIRM_MTR			BIT(1)
+#define GWMTIRM_MTR		BIT(1)
 
 #define GWVCC_VEM_SC_TAG	(0x3 << 16)
 
 #define GWARIRM_ARIOG		BIT(0)
-#define GWARIRM_ARR			BIT(1)
+#define GWARIRM_ARR		BIT(1)
 
 #define GWMDNC_TSDMN(num)	(((num) << 16) & GENMASK(17, 16))
 #define GWMDNC_TXDMN(num)	(((num) << 8) & GENMASK(12, 8))
 #define GWMDNC_RXDMN(num)	((num) & GENMASK(4, 0))
 
-#define GWDCC_BALR			BIT(24)
+#define GWDCC_BALR		BIT(24)
 #define GWDCC_DCP_MASK		GENMASK(18, 16)
 #define GWDCC_DCP(prio)		FIELD_PREP(GWDCC_DCP_MASK, (prio))
-#define GWDCC_DQT			BIT(11)
-#define GWDCC_ETS			BIT(9)
-#define GWDCC_EDE			BIT(8)
+#define GWDCC_DQT		BIT(11)
+#define GWDCC_ETS		BIT(9)
+#define GWDCC_EDE		BIT(8)
+#define GWCKSC_USMFSPE		BIT(31)
 
 #define GWTRC(queue)		(GWTRC0 + (queue) / 32 * 4)
 #define GWTPC_PPPL(ipv)		BIT(ipv)
 #define GWDCC_OFFS(queue)	(GWDCC0 + (queue) * 4)
 
-#define GWDIS(i)			(GWDIS0 + (i) * 0x10)
-#define GWDIE(i)			(GWDIE0 + (i) * 0x10)
-#define GWDID(i)			(GWDID0 + (i) * 0x10)
+#define GWDIS(i)		(GWDIS0 + (i) * 0x10)
+#define GWDIE(i)		(GWDIE0 + (i) * 0x10)
+#define GWDID(i)		(GWDID0 + (i) * 0x10)
 
 /* COMA */
-#define RRC_RR				BIT(0)
-#define RRC_RR_CLR			0
+#define RRC_RR			BIT(0)
+#define RRC_RR_CLR		0
 #define	RCEC_ACE_DEFAULT	(BIT(0) | BIT(AGENT_INDEX_GWCA))
-#define RCEC_RCE			BIT(16)
-#define RCDC_RCD			BIT(16)
+#define RCEC_RCE		BIT(16)
+#define RCDC_RCD		BIT(16)
 
 #define CABPIRM_BPIOG		BIT(0)
-#define CABPIRM_BPR			BIT(1)
+#define CABPIRM_BPR		BIT(1)
 
 #define CABPPFLC_INIT_VALUE	0x00800080
 
 /* MFWD */
-#define FWPC0_LTHTA			BIT(0)
-#define FWPC0_IP4UE			BIT(3)
-#define FWPC0_IP4TE			BIT(4)
-#define FWPC0_IP4OE			BIT(5)
-#define FWPC0_L2SE			BIT(9)
-#define FWPC0_IP4EA			BIT(10)
-#define FWPC0_IPDSA			BIT(12)
-#define FWPC0_IPHLA			BIT(18)
+#define FWPC0_LTHTA		BIT(0)
+#define FWPC0_IP4UE		BIT(3)
+#define FWPC0_IP4TE		BIT(4)
+#define FWPC0_IP4OE		BIT(5)
+#define FWPC0_L2SE		BIT(9)
+#define FWPC0_IP4EA		BIT(10)
+#define FWPC0_IPDSA		BIT(12)
+#define FWPC0_IPHLA		BIT(18)
 #define FWPC0_MACSDA		BIT(20)
 #define FWPC0_MACHLA		BIT(26)
 #define FWPC0_MACHMA		BIT(27)
 #define FWPC0_VLANSA		BIT(28)
 
-#define FWPC0(i)			(FWPC00 + (i) * 0x10)
+#define FWPC0(i)		(FWPC00 + (i) * 0x10)
 #define FWPC0_DEFAULT		(FWPC0_LTHTA | FWPC0_IP4UE | FWPC0_IP4TE | \
 				 FWPC0_IP4OE | FWPC0_L2SE | FWPC0_IP4EA | \
 				 FWPC0_IPDSA | FWPC0_IPHLA | FWPC0_MACSDA | \
 				 FWPC0_MACHLA |	FWPC0_MACHMA | FWPC0_VLANSA)
-#define FWPC1(i)			(FWPC10 + (i) * 0x10)
-#define FWPC1_DDE			BIT(0)
+#define FWPC1(i)		(FWPC10 + (i) * 0x10)
+#define FWPC1_DDE		BIT(0)
 
-#define	FWPBFC(i)			(FWPBFC00 + (i) * 0x10)
+#define	FWPBFC(i)		(FWPBFC00 + (i) * 0x10)
 
 #define FWPBFCSDC(j, i)		(FWPBFCSDC00 + (i) * 0x20 + (j) * 0x04)
 
 /* TOP */
-#define  TPDEMIMC0(queue)		(TPDEMIMC0 + (queue) * 4)
+#define  TPDEMIMC0(queue)	(TPDEMIMC0 + (queue) * 4)
 
 /* Descriptors */
 enum RX_DS_CC_BIT {
@@ -957,15 +954,15 @@ enum DIE_DT {
 	DT_EOS			= 0xf0,
 	/* HW/SW arbitration */
 	DT_FEMPTY		= 0x40,
-	DT_FEMPTY_IS	= 0x10,
-	DT_FEMPTY_IC	= 0x20,
-	DT_FEMPTY_ND	= 0x30,
-	DT_FEMPTY_START	= 0x50,
-	DT_FEMPTY_MID	= 0x60,
-	DT_FEMPTY_END	= 0x70,
+	DT_FEMPTY_IS		= 0x10,
+	DT_FEMPTY_IC		= 0x20,
+	DT_FEMPTY_ND		= 0x30,
+	DT_FEMPTY_START		= 0x50,
+	DT_FEMPTY_MID		= 0x60,
+	DT_FEMPTY_END		= 0x70,
 
 	DT_MASK			= 0xf0,
-	DIE				= 0x08,	/* Descriptor Interrupt Enable */
+	DIE			= 0x08,	/* Descriptor Interrupt Enable */
 };
 
 /* Both transmission and reception */
@@ -982,32 +979,32 @@ enum DIE_DT {
 /* For reception */
 #define INFO1_SPN(port)		((u64)(port) << 36ULL)
 
-struct rswitch_desc {
+struct rsw3_desc {
 	__le16 info_ds;		/* Descriptor size */
-	u8 die_dt;			/* Descriptor interrupt enable and type */
+	u8 die_dt;		/* Descriptor interrupt enable and type */
 	__u8  dptrh;		/* Descriptor pointer MSB */
 	__le32 dptrl;		/* Descriptor pointer LSW */
 } __packed;
 
-struct rswitch_ts_desc {
-	struct rswitch_desc desc;
+struct rsw3_ts_desc {
+	struct rsw3_desc desc;
 	__le32 ts_nsec;
 	__le32 ts_sec;
 } __packed;
 
-struct rswitch_ext_desc {
-	struct rswitch_desc desc;
+struct rsw3_ext_desc {
+	struct rsw3_desc desc;
 	__le64 info1;
 } __packed;
 
-struct rswitch_ext_ts_desc {
-	struct rswitch_desc desc;
+struct rsw3_ext_ts_desc {
+	struct rsw3_desc desc;
 	__le64 info1;
 	__le32 ts_nsec;
 	__le32 ts_sec;
 } __packed;
 
-struct rswitch_etha {
+struct rsw3_etha {
 	unsigned int index;
 	void __iomem *addr;
 	void __iomem *coma_addr;
@@ -1023,17 +1020,18 @@ struct rswitch_etha {
 	 * this flag to avoid multiple initialization.
 	 */
 	bool operated;
+	bool connect_to_xpcs;
 };
 
 /* The datasheet said descriptor "chain" and/or "queue". For consistency of
  * name, this driver calls "queue".
  */
-struct rswitch_gwca_queue {
+struct rsw3_gwca_queue {
 	int index;
 	bool dir_tx;
 	union {
-		struct rswitch_ext_desc *tx_ring;
-		struct rswitch_ext_ts_desc *rx_ring;
+		struct rsw3_ext_desc *tx_ring;
+		struct rsw3_ext_ts_desc *rx_ring;
 	};
 
 	dma_addr_t ring_dma;
@@ -1059,12 +1057,12 @@ struct rswitch_gwca_queue {
 };
 
 #define RSWITCH3_NUM_IRQ_REGS	(RSWITCH3_MAX_NUM_QUEUES / BITS_PER_TYPE(u32))
-struct rswitch_gwca {
+struct rsw3_gwca {
 	unsigned int index;
-	struct rswitch_desc *linkfix_table;
+	struct rsw3_desc *linkfix_table;
 	dma_addr_t linkfix_table_dma;
 	u32 linkfix_table_size;
-	struct rswitch_gwca_queue *queues;
+	struct rsw3_gwca_queue *queues;
 	int num_queues;
 	DECLARE_BITMAP(used, RSWITCH3_MAX_NUM_QUEUES);
 	u32 tx_irq_bits[RSWITCH3_NUM_IRQ_REGS];
@@ -1073,52 +1071,52 @@ struct rswitch_gwca {
 };
 
 #define NUM_QUEUES_PER_NDEV	2
-struct rswitch_device {
-	struct rswitch_private *priv;
+struct rsw3_device {
+	struct rsw3_private *priv;
 	struct net_device *ndev;
 	struct napi_struct napi;
 	void __iomem *addr;
-	struct rswitch_gwca_queue *tx_queue;
-	struct rswitch_gwca_queue *rx_queue;
+	struct rsw3_gwca_queue *tx_queue;
+	struct rsw3_gwca_queue *rx_queue;
 	u8 ts_tag;
 	bool disabled;
 
 	int port;
-	struct rswitch_etha *etha;
+	int etha_mii_idx;
+	struct rsw3_etha *etha;
+	struct rsw3_etha *etha_mii;
 	struct device_node *np_port;
-	struct phy *serdes;
+	struct phy *pcs;
 };
 
-struct rswitch_mfwd_mac_table_entry {
+struct rsw3_mfwd_mac_table_entry {
 	int queue_index;
 	unsigned char addr[MAX_ADDR_LEN];
 };
 
-struct rswitch_mfwd {
-	struct rswitch_mac_table_entry *mac_table_entries;
+struct rsw3_mfwd {
+	struct rsw3_mac_table_entry *mac_table_entries;
 	int num_mac_table_entries;
 };
 
-struct rswitch_private {
+struct rsw3_private {
 	struct platform_device *pdev;
 	void __iomem *addr;
 	struct rcar_gen4_ptp_private *ptp_priv;
 
-	struct rswitch_device *rdev[RSWITCH3_NUM_PORTS];
+	struct rsw3_device *rdev[RSWITCH3_NUM_PORTS];
 	DECLARE_BITMAP(opened_ports, RSWITCH3_NUM_PORTS);
 
-	struct rswitch_gwca gwca;
-	struct rswitch_etha etha[RSWITCH3_NUM_PORTS];
-	struct rswitch_mfwd mfwd;
+	struct rsw3_gwca gwca;
+	struct rsw3_etha etha[RSWITCH3_NUM_PORTS];
+	struct rsw3_etha etha_mii[RSWITCH3_NUM_PORTS];
+	struct rsw3_mfwd mfwd;
 
 	spinlock_t lock;	/* lock interrupt registers' control */
 	struct clk *clk;
 
-	bool etha_no_runtime_change;
 	bool gwca_halt;
 
-	/* Parameter for VPF environment which can config in dts file */
-	bool	vpf_mode;
 	struct clk *rsw_clk;
 	struct clk *phy_clk;
 };
-- 
2.34.1

